// ng-primitives/fesm2022/ng-primitives-button.mjs
import * as i0 from '@angular/core';
import { input, booleanAttribute, Directive } from '@angular/core';
import { createStateToken, createStateProvider, createStateInjector, createState } from 'ng-primitives/state';
import { setupInteractions } from 'ng-primitives/interactions';
import { injectElementRef } from 'ng-primitives/internal';
import { booleanAttributeBinding } from 'ng-primitives/utils';

/**
 * The state token  for the Button primitive.
 */
const NgpButtonStateToken = createStateToken('Button');
/**
 * Provides the Button state.
 */
const provideButtonState = createStateProvider(NgpButtonStateToken);
/**
 * Injects the Button state.
 */
const injectButtonState = createStateInjector(NgpButtonStateToken);
/**
 * The Button state registration function.
 */
const buttonState = createState(NgpButtonStateToken);

/**
 * @internal
 * Setup the button interactions and attributes.
 * @param options - The options for the button.
 */
function setupButton({ disabled }) {
    const elementRef = injectElementRef();
    const isButton = elementRef.nativeElement.tagName.toLowerCase() === 'button';
    setupInteractions({ hover: true, press: true, focusVisible: true, disabled });
    // add the `data-disabled` attribute to the element
    booleanAttributeBinding(elementRef.nativeElement, 'data-disabled', disabled);
    // add the `disabled` attribute to the element if it is a button
    if (isButton) {
        booleanAttributeBinding(elementRef.nativeElement, 'disabled', disabled);
    }
}

class NgpButton {
    constructor() {
        /**
         * Whether the button is disabled.
         */
        this.disabled = input(false, {
            transform: booleanAttribute,
        });
        /**
         * The button state.
         */
        this.state = buttonState(this);
        setupButton({ disabled: this.state.disabled });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpButton, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "19.2.11", type: NgpButton, isStandalone: true, selector: "[ngpButton]", inputs: { disabled: { classPropertyName: "disabled", publicName: "disabled", isSignal: true, isRequired: false, transformFunction: null } }, providers: [provideButtonState({ inherit: false })], exportAs: ["ngpButton"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpButton, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpButton]',
                    exportAs: 'ngpButton',
                    providers: [provideButtonState({ inherit: false })],
                }]
        }], ctorParameters: () => [] });

/**
 * Generated bundle index. Do not edit.
 */

export { NgpButton, injectButtonState, provideButtonState, setupButton };
