// ng-primitives/fesm2022/ng-primitives-switch.mjs
import * as i0 from '@angular/core';
import { Directive, input, booleanAttribute, output, HostListener } from '@angular/core';
import { setupInteractions } from 'ng-primitives/interactions';
import { createStateToken, createStateProvider, createStateInjector, createState } from 'ng-primitives/state';
import { setupFormControl } from 'ng-primitives/form-field';
import { injectElementRef } from 'ng-primitives/internal';
import { uniqueId } from 'ng-primitives/utils';

/**
 * The state token  for the Switch primitive.
 */
const NgpSwitchStateToken = createStateToken('Switch');
/**
 * Provides the Switch state.
 */
const provideSwitchState = createStateProvider(NgpSwitchStateToken);
/**
 * Injects the Switch state.
 */
const injectSwitchState = createStateInjector(NgpSwitchStateToken);
/**
 * The Switch state registration function.
 */
const switchState = createState(NgpSwitchStateToken);

/**
 * Apply the `ngpSwitchThumb` directive to an element within a switch to represent the thumb.
 */
class NgpSwitchThumb {
    constructor() {
        /**
         * Access the switch state.
         */
        this.state = injectSwitchState();
        setupInteractions({
            hover: true,
            focusVisible: true,
            press: true,
            disabled: this.state().disabled,
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpSwitchThumb, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "19.2.11", type: NgpSwitchThumb, isStandalone: true, selector: "[ngpSwitchThumb]", host: { properties: { "attr.data-checked": "state().checked() ? \"\" : null", "attr.data-disabled": "state().disabled() ? \"\" : null" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpSwitchThumb, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpSwitchThumb]',
                    host: {
                        '[attr.data-checked]': 'state().checked() ? "" : null',
                        '[attr.data-disabled]': 'state().disabled() ? "" : null',
                    },
                }]
        }], ctorParameters: () => [] });

/**
 * Apply the `ngpSwitch` directive to an element to manage the checked state.
 */
class NgpSwitch {
    constructor() {
        /**
         * Access the element ref.
         */
        this.elementRef = injectElementRef();
        /**
         * Determine if the switch is a button
         */
        this.isButton = this.elementRef.nativeElement.tagName === 'BUTTON';
        /**
         * The id of the switch. If not provided, a unique id will be generated.
         */
        this.id = input(uniqueId('ngp-switch'));
        /**
         * Determine if the switch is checked.
         * @default false
         */
        this.checked = input(false, {
            alias: 'ngpSwitchChecked',
            transform: booleanAttribute,
        });
        /**
         * Emits when the checked state changes.
         */
        this.checkedChange = output({
            alias: 'ngpSwitchCheckedChange',
        });
        /**
         * Determine if the switch is disabled.
         * @default false
         */
        this.disabled = input(false, {
            alias: 'ngpSwitchDisabled',
            transform: booleanAttribute,
        });
        /**
         * The switch state.
         * @internal
         */
        this.state = switchState(this);
        setupInteractions({
            hover: true,
            press: true,
            focusVisible: true,
            disabled: this.state.disabled,
        });
        setupFormControl({ id: this.state.id, disabled: this.state.disabled });
    }
    /**
     * Toggle the checked state.
     */
    toggle() {
        if (this.state.disabled()) {
            return;
        }
        const checked = !this.state.checked();
        this.state.checked.set(checked);
        this.checkedChange.emit(checked);
    }
    /**
     * Handle the keydown event.
     */
    onKeyDown(event) {
        // Prevent the default action of the space key, which is to scroll the page.
        event.preventDefault();
        // If the switch is not a button then the space key will not toggle the checked state automatically,
        // so we need to do it manually.
        if (!this.isButton) {
            this.toggle();
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpSwitch, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "19.2.11", type: NgpSwitch, isStandalone: true, selector: "[ngpSwitch]", inputs: { id: { classPropertyName: "id", publicName: "id", isSignal: true, isRequired: false, transformFunction: null }, checked: { classPropertyName: "checked", publicName: "ngpSwitchChecked", isSignal: true, isRequired: false, transformFunction: null }, disabled: { classPropertyName: "disabled", publicName: "ngpSwitchDisabled", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { checkedChange: "ngpSwitchCheckedChange" }, host: { attributes: { "role": "switch" }, listeners: { "click": "toggle()", "keydown.space": "onKeyDown($event)" }, properties: { "id": "id()", "attr.type": "isButton ? \"button\" : null", "attr.aria-checked": "state.checked()", "attr.data-checked": "state.checked() ? \"\" : null", "attr.disabled": "isButton && state.disabled() ? \"\" : null", "attr.data-disabled": "state.disabled() ? \"\" : null", "attr.aria-disabled": "state.disabled()", "attr.tabindex": "state.disabled() ? -1 : 0" } }, providers: [provideSwitchState()], exportAs: ["ngpSwitch"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpSwitch, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpSwitch]',
                    exportAs: 'ngpSwitch',
                    providers: [provideSwitchState()],
                    host: {
                        role: 'switch',
                        '[id]': 'id()',
                        '[attr.type]': 'isButton ? "button" : null',
                        '[attr.aria-checked]': 'state.checked()',
                        '[attr.data-checked]': 'state.checked() ? "" : null',
                        '[attr.disabled]': 'isButton && state.disabled() ? "" : null',
                        '[attr.data-disabled]': 'state.disabled() ? "" : null',
                        '[attr.aria-disabled]': 'state.disabled()',
                        '[attr.tabindex]': 'state.disabled() ? -1 : 0',
                    },
                }]
        }], ctorParameters: () => [], propDecorators: { toggle: [{
                type: HostListener,
                args: ['click']
            }], onKeyDown: [{
                type: HostListener,
                args: ['keydown.space', ['$event']]
            }] } });

/**
 * Generated bundle index. Do not edit.
 */

export { NgpSwitch, NgpSwitchThumb, injectSwitchState, provideSwitchState };
