// ng-primitives/fesm2022/ng-primitives-toggle.mjs
import * as i0 from '@angular/core';
import { inject, ElementRef, input, booleanAttribute, output, HostListener, Directive } from '@angular/core';
import { createStateToken, createStateProvider, createStateInjector, createState } from 'ng-primitives/state';

/**
 * The state token  for the Toggle primitive.
 */
const NgpToggleStateToken = createStateToken('Toggle');
/**
 * Provides the Toggle state.
 */
const provideToggleState = createStateProvider(NgpToggleStateToken);
/**
 * Injects the Toggle state.
 */
const injectToggleState = createStateInjector(NgpToggleStateToken);
/**
 * The Toggle state registration function.
 */
const toggleState = createState(NgpToggleStateToken);

/**
 * Apply the `ngpToggle` directive to an element to manage the toggle state. This must be applied to a `button` element.
 */
class NgpToggle {
    constructor() {
        /**
         * Access the element.
         */
        this.element = inject(ElementRef);
        /**
         * Whether the toggle is selected.
         * @default false
         */
        this.selected = input(false, {
            alias: 'ngpToggleSelected',
            transform: booleanAttribute,
        });
        /**
         * Emits when the selected state changes.
         */
        this.selectedChange = output({
            alias: 'ngpToggleSelectedChange',
        });
        /**
         * Whether the toggle is disabled.
         * @default false
         */
        this.disabled = input(false, {
            alias: 'ngpToggleDisabled',
            transform: booleanAttribute,
        });
        /**
         * Determine if the element is a button.
         */
        this.isButton = this.element.nativeElement.tagName === 'BUTTON';
        /**
         * The state for the toggle primitive.
         * @internal
         */
        this.state = toggleState(this);
    }
    /**
     * Toggle the selected state.
     */
    toggle() {
        if (this.state.disabled()) {
            return;
        }
        const isSelected = !this.state.selected();
        this.state.selected.set(isSelected);
        this.selectedChange.emit(isSelected);
    }
    /**
     * If the element is not a button or a link the space key should toggle the selected state.
     */
    onKeyDown(event) {
        if (!this.isButton && this.element.nativeElement.tagName !== 'A') {
            event.preventDefault();
            this.toggle();
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpToggle, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "19.2.11", type: NgpToggle, isStandalone: true, selector: "[ngpToggle]", inputs: { selected: { classPropertyName: "selected", publicName: "ngpToggleSelected", isSignal: true, isRequired: false, transformFunction: null }, disabled: { classPropertyName: "disabled", publicName: "ngpToggleDisabled", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { selectedChange: "ngpToggleSelectedChange" }, host: { listeners: { "click": "toggle()", "keydown.space": "onKeyDown($event)" }, properties: { "attr.type": "isButton ? \"button\" : null", "attr.aria-pressed": "state.selected()", "attr.data-selected": "state.selected() ? \"\" : null", "attr.data-disabled": "state.disabled() ? \"\" : null" } }, providers: [provideToggleState()], exportAs: ["ngpToggle"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpToggle, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpToggle]',
                    exportAs: 'ngpToggle',
                    providers: [provideToggleState()],
                    host: {
                        '[attr.type]': 'isButton ? "button" : null',
                        '[attr.aria-pressed]': 'state.selected()',
                        '[attr.data-selected]': 'state.selected() ? "" : null',
                        '[attr.data-disabled]': 'state.disabled() ? "" : null',
                    },
                }]
        }], propDecorators: { toggle: [{
                type: HostListener,
                args: ['click']
            }], onKeyDown: [{
                type: HostListener,
                args: ['keydown.space', ['$event']]
            }] } });

/**
 * Generated bundle index. Do not edit.
 */

export { NgpToggle, injectToggleState, provideToggleState };
