// ng-primitives/fesm2022/ng-primitives-input.mjs
import * as i0 from '@angular/core';
import { input, booleanAttribute, Directive } from '@angular/core';
import * as i1 from 'ng-primitives/autofill';
import { NgpAutofill } from 'ng-primitives/autofill';
import { setupFormControl } from 'ng-primitives/form-field';
import { setupInteractions } from 'ng-primitives/interactions';
import { injectElementRef } from 'ng-primitives/internal';
import { injectSearchState } from 'ng-primitives/search';
import { uniqueId } from 'ng-primitives/utils';
import { createStateToken, createStateProvider, createStateInjector, createState } from 'ng-primitives/state';

/**
 * The state token  for the Input primitive.
 */
const NgpInputStateToken = createStateToken('Input');
/**
 * Provides the Input state.
 */
const provideInputState = createStateProvider(NgpInputStateToken);
/**
 * Injects the Input state.
 */
const injectInputState = createStateInjector(NgpInputStateToken);
/**
 * The Input state registration function.
 */
const inputState = createState(NgpInputStateToken);

class NgpInput {
    constructor() {
        /**
         * The id of the input.
         */
        this.id = input(uniqueId('ngp-input'));
        /**
         * The input may be used within a search field, if so we need to register it.
         */
        this.searchState = injectSearchState({ optional: true });
        /**
         * Access the element reference.
         */
        this.elementRef = injectElementRef();
        /**
         * Whether the element is disabled.
         */
        this.disabled = input(false, {
            transform: booleanAttribute,
        });
        /**
         * The input state.
         */
        this.state = inputState(this);
        setupInteractions({
            hover: true,
            press: true,
            focus: true,
            disabled: this.state.disabled,
        });
        // Set up the form control with the id and disabled state.
        this.status = setupFormControl({ id: this.state.id, disabled: this.state.disabled });
        this.searchState()?.registerInput(this.elementRef.nativeElement);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpInput, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "19.2.11", type: NgpInput, isStandalone: true, selector: "input[ngpInput]", inputs: { id: { classPropertyName: "id", publicName: "id", isSignal: true, isRequired: false, transformFunction: null }, disabled: { classPropertyName: "disabled", publicName: "disabled", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "attr.id": "id()", "attr.disabled": "status().disabled ? \"\" : null" } }, providers: [provideInputState()], exportAs: ["ngpInput"], hostDirectives: [{ directive: i1.NgpAutofill }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpInput, decorators: [{
            type: Directive,
            args: [{
                    selector: 'input[ngpInput]',
                    exportAs: 'ngpInput',
                    providers: [provideInputState()],
                    hostDirectives: [NgpAutofill],
                    host: {
                        '[attr.id]': 'id()',
                        '[attr.disabled]': 'status().disabled ? "" : null',
                    },
                }]
        }], ctorParameters: () => [] });

/**
 * Generated bundle index. Do not edit.
 */

export { NgpInput, injectInputState, provideInputState };
