// ng-primitives/fesm2022/ng-primitives-textarea.mjs
import * as i0 from '@angular/core';
import { input, booleanAttribute, Directive } from '@angular/core';
import { setupFormControl } from 'ng-primitives/form-field';
import { setupInteractions } from 'ng-primitives/interactions';
import { uniqueId } from 'ng-primitives/utils';
import { createStateToken, createStateProvider, createStateInjector, createState } from 'ng-primitives/state';

/**
 * The state token  for the Textarea primitive.
 */
const NgpTextareaStateToken = createStateToken('Textarea');
/**
 * Provides the Textarea state.
 */
const provideTextareaState = createStateProvider(NgpTextareaStateToken);
/**
 * Injects the Textarea state.
 */
const injectTextareaState = createStateInjector(NgpTextareaStateToken);
/**
 * The Textarea state registration function.
 */
const textareaState = createState(NgpTextareaStateToken);

class NgpTextarea {
    constructor() {
        /**
         * The id of the textarea. If not provided, a unique id will be generated.
         */
        this.id = input(uniqueId('ngp-textarea'));
        /**
         * Whether the element is disabled.
         */
        this.disabled = input(false, {
            transform: booleanAttribute,
        });
        /**
         * The state of the textarea.
         */
        this.state = textareaState(this);
        setupInteractions({
            hover: true,
            press: true,
            focus: true,
            disabled: this.state.disabled,
        });
        setupFormControl({ id: this.state.id, disabled: this.state.disabled });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpTextarea, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "19.2.11", type: NgpTextarea, isStandalone: true, selector: "[ngpTextarea]", inputs: { id: { classPropertyName: "id", publicName: "id", isSignal: true, isRequired: false, transformFunction: null }, disabled: { classPropertyName: "disabled", publicName: "disabled", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "id": "id()", "attr.disabled": "disabled() ? \"\" : null" } }, providers: [provideTextareaState()], exportAs: ["ngpTextarea"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpTextarea, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpTextarea]',
                    exportAs: 'ngpTextarea',
                    providers: [provideTextareaState()],
                    host: {
                        '[id]': 'id()',
                        '[attr.disabled]': 'disabled() ? "" : null',
                    },
                }]
        }], ctorParameters: () => [] });

/**
 * Generated bundle index. Do not edit.
 */

export { NgpTextarea, injectTextareaState, provideTextareaState };
