// ng-primitives/fesm2022/ng-primitives-tabs.mjs
import * as i0 from '@angular/core';
import { InjectionToken, inject, HOST_TAG_NAME, input, booleanAttribute, computed, HostListener, Directive, output, signal } from '@angular/core';
import { setupInteractions } from 'ng-primitives/interactions';
import * as i1 from 'ng-primitives/roving-focus';
import { NgpRovingFocusItem, injectRovingFocusGroupState, NgpRovingFocusGroup } from 'ng-primitives/roving-focus';
import { createStateToken, createStateProvider, createStateInjector, createState } from 'ng-primitives/state';
import { explicitEffect } from 'ng-primitives/internal';
import { uniqueId } from 'ng-primitives/utils';

const defaultTabsConfig = {
    orientation: 'horizontal',
    activateOnFocus: true,
    wrap: true,
};
const NgpTabsConfigToken = new InjectionToken('NgpTabsConfigToken');
/**
 * Provide the default Tabs configuration
 * @param config The Tabs configuration
 * @returns The provider
 */
function provideTabsConfig(config) {
    return [
        {
            provide: NgpTabsConfigToken,
            useValue: { ...defaultTabsConfig, ...config },
        },
    ];
}
/**
 * Inject the Tabs configuration
 * @returns The global Tabs configuration
 */
function injectTabsConfig() {
    return inject(NgpTabsConfigToken, { optional: true }) ?? defaultTabsConfig;
}

/**
 * The state token  for the Tabset primitive.
 */
const NgpTabsetStateToken = createStateToken('Tabset');
/**
 * Provides the Tabset state.
 */
const provideTabsetState = createStateProvider(NgpTabsetStateToken);
/**
 * Injects the Tabset state.
 */
const injectTabsetState = createStateInjector(NgpTabsetStateToken);
/**
 * The Tabset state registration function.
 */
const tabsetState = createState(NgpTabsetStateToken);

/**
 * Apply the `ngpTabButton` directive to an element within a tab list to represent a tab button. This directive should be applied to a button element.
 */
class NgpTabButton {
    constructor() {
        /**
         * Access the tag host name
         */
        this.tagName = inject(HOST_TAG_NAME);
        /**
         * Access the tabset state
         */
        this.state = injectTabsetState();
        /**
         * The value of the tab this trigger controls
         * @required
         */
        this.value = input(undefined, { alias: 'ngpTabButtonValue' });
        /**
         * Whether the tab is disabled
         * @default false
         */
        this.disabled = input(false, {
            alias: 'ngpTabButtonDisabled',
            transform: booleanAttribute,
        });
        /**
         * Determine the id of the tab button
         * @internal
         */
        this.id = input();
        /**
         * Determine a unique id for the tab button if not provided
         * @internal
         */
        this.buttonId = computed(() => this.id() ?? `${this.state().id()}-button-${this.value()}`);
        /**
         * Determine the aria-controls of the tab button
         * @internal
         */
        this.ariaControls = computed(() => `${this.state().id()}-panel-${this.value()}`);
        /**
         * Whether the tab is active
         */
        this.active = computed(() => this.state().selectedTab() === this.value());
        this.state().registerTab(this);
        setupInteractions({
            hover: true,
            press: true,
            focusVisible: true,
            disabled: this.disabled,
        });
    }
    ngOnInit() {
        if (this.value() === undefined) {
            throw new Error('ngpTabButton: value is required');
        }
    }
    ngOnDestroy() {
        this.state().unregisterTab(this);
    }
    /**
     * Select the tab this trigger controls
     */
    select() {
        if (this.disabled() === false) {
            this.state().select(this.value());
        }
    }
    /**
     * On focus select the tab this trigger controls if activateOnFocus is true
     */
    activateOnFocus() {
        if (this.state().activateOnFocus()) {
            this.select();
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpTabButton, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "19.2.11", type: NgpTabButton, isStandalone: true, selector: "[ngpTabButton]", inputs: { value: { classPropertyName: "value", publicName: "ngpTabButtonValue", isSignal: true, isRequired: false, transformFunction: null }, disabled: { classPropertyName: "disabled", publicName: "ngpTabButtonDisabled", isSignal: true, isRequired: false, transformFunction: null }, id: { classPropertyName: "id", publicName: "id", isSignal: true, isRequired: false, transformFunction: null } }, host: { attributes: { "role": "tab" }, listeners: { "click": "select()", "focus": "activateOnFocus()" }, properties: { "attr.id": "buttonId()", "attr.aria-controls": "ariaControls()", "attr.data-active": "active() ? \"\" : null", "attr.data-disabled": "disabled() ? \"\" : null", "attr.disabled": "tagName === \"button\" && disabled() ? \"\" : null", "attr.data-orientation": "state().orientation()" } }, exportAs: ["ngpTabButton"], hostDirectives: [{ directive: i1.NgpRovingFocusItem, inputs: ["ngpRovingFocusItemDisabled", "ngpTabButtonDisabled"] }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpTabButton, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpTabButton]',
                    exportAs: 'ngpTabButton',
                    host: {
                        role: 'tab',
                        '[attr.id]': 'buttonId()',
                        '[attr.aria-controls]': 'ariaControls()',
                        '[attr.data-active]': 'active() ? "" : null',
                        '[attr.data-disabled]': 'disabled() ? "" : null',
                        '[attr.disabled]': 'tagName === "button" && disabled() ? "" : null',
                        '[attr.data-orientation]': 'state().orientation()',
                    },
                    hostDirectives: [
                        {
                            directive: NgpRovingFocusItem,
                            inputs: ['ngpRovingFocusItemDisabled: ngpTabButtonDisabled'],
                        },
                    ],
                }]
        }], ctorParameters: () => [], propDecorators: { select: [{
                type: HostListener,
                args: ['click']
            }], activateOnFocus: [{
                type: HostListener,
                args: ['focus']
            }] } });

/**
 * Apply the `ngpTabList` directive to an element that represents the list of tab buttons.
 */
class NgpTabList {
    constructor() {
        /**
         * Access the tabset state
         */
        this.state = injectTabsetState();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpTabList, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "19.2.11", type: NgpTabList, isStandalone: true, selector: "[ngpTabList]", host: { attributes: { "role": "tablist" }, properties: { "attr.aria-orientation": "state().orientation()", "attr.data-orientation": "state().orientation()" } }, exportAs: ["ngpTabList"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpTabList, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpTabList]',
                    exportAs: 'ngpTabList',
                    host: {
                        role: 'tablist',
                        '[attr.aria-orientation]': 'state().orientation()',
                        '[attr.data-orientation]': 'state().orientation()',
                    },
                }]
        }] });

const NgpTabPanelToken = new InjectionToken('NgpTabPanelToken');
/**
 * Inject the TabPanel directive instance
 * @returns The TabPanel directive instance
 */
function injectTabPanel() {
    return inject(NgpTabPanelToken);
}

/**
 * Apply the `ngpTabPanel` directive to an element that represents the content of a tab.
 */
class NgpTabPanel {
    constructor() {
        /**
         * Access the tabset
         */
        this.state = injectTabsetState();
        /**
         * The value of the tab
         * @required
         */
        this.value = input(undefined, { alias: 'ngpTabPanelValue' });
        /**
         * Determine the id of the tab panel
         * @internal
         */
        this.id = input();
        /**
         * Determine a unique id for the tab panel if not provided
         * @internal
         */
        this.panelId = computed(() => this.id() ?? `${this.state().id()}-panel-${this.value()}`);
        /**
         * Determine the aria-labelledby of the tab panel
         * @internal
         */
        this.labelledBy = computed(() => `${this.state().id()}-button-${this.value()}`);
        /**
         * Whether the tab is active
         */
        this.active = computed(() => this.state().selectedTab() === this.value());
    }
    ngOnInit() {
        if (this.value() === undefined) {
            throw new Error('ngpTabPanel: value is required');
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpTabPanel, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "19.2.11", type: NgpTabPanel, isStandalone: true, selector: "[ngpTabPanel]", inputs: { value: { classPropertyName: "value", publicName: "ngpTabPanelValue", isSignal: true, isRequired: false, transformFunction: null }, id: { classPropertyName: "id", publicName: "id", isSignal: true, isRequired: false, transformFunction: null } }, host: { attributes: { "role": "tabpanel", "tabIndex": "0" }, properties: { "id": "panelId()", "attr.aria-labelledby": "labelledBy()", "attr.data-active": "active() ? \"\" : null", "attr.data-orientation": "state().orientation()" } }, providers: [{ provide: NgpTabPanelToken, useExisting: NgpTabPanel }], exportAs: ["ngpTabPanel"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpTabPanel, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpTabPanel]',
                    exportAs: 'ngpTabPanel',
                    providers: [{ provide: NgpTabPanelToken, useExisting: NgpTabPanel }],
                    host: {
                        role: 'tabpanel',
                        tabIndex: '0',
                        '[id]': 'panelId()',
                        '[attr.aria-labelledby]': 'labelledBy()',
                        '[attr.data-active]': 'active() ? "" : null',
                        '[attr.data-orientation]': 'state().orientation()',
                    },
                }]
        }] });

/**
 * Apply the `ngpTabset` directive to an element to manage the tabs.
 */
class NgpTabset {
    constructor() {
        /**
         * Access the global tabset configuration
         */
        this.config = injectTabsConfig();
        /**
         * Access the roving focus group state
         */
        this.rovingFocusGroupState = injectRovingFocusGroupState();
        /**
         * Define the id for the tabset
         */
        this.id = input(uniqueId('ngp-tabset'));
        /**
         * Define the active tab
         */
        this.value = input(undefined, {
            alias: 'ngpTabsetValue',
        });
        /**
         * Emit the value of the selected tab when it changes
         */
        this.valueChange = output({
            alias: 'ngpTabsetValueChange',
        });
        /**
         * The orientation of the tabset
         * @default 'horizontal'
         */
        this.orientation = input(this.config.orientation, {
            alias: 'ngpTabsetOrientation',
        });
        /**
         * Whether tabs should activate on focus
         */
        this.activateOnFocus = input(this.config.activateOnFocus, {
            alias: 'ngpTabsetActivateOnFocus',
            transform: booleanAttribute,
        });
        /**
         * Access the tabs within the tabset
         * @internal
         */
        this.buttons = signal([]);
        /**
         * @internal
         * Get the id of the selected tab
         */
        this.selectedTab = computed(() => {
            const buttons = this.buttons();
            // if there are no tabs then return the selected value
            // if there is a value set and a tab with that value exists, return the value
            if (buttons.length === 0 || buttons.some(button => button.value() === this.state.value())) {
                return this.state.value();
            }
            // otherwise return the first non-disabled tab's value
            return buttons.find(button => !button.disabled())?.value();
        });
        /**
         * The state of the tabset
         */
        this.state = tabsetState(this);
        explicitEffect([this.state.orientation], ([orientation]) => this.rovingFocusGroupState().orientation.set(orientation));
    }
    /**
     * Select a tab by its value
     * @param value The value of the tab to select
     */
    select(value) {
        // if the value is already selected, do nothing
        if (this.state.value() === value) {
            return;
        }
        this.state.value.set(value);
        this.valueChange.emit(value);
    }
    /**
     * @internal
     * Register a tab with the tabset
     */
    registerTab(tab) {
        this.buttons.update(buttons => [...buttons, tab]);
    }
    /**
     * @internal
     * Unregister a tab with the tabset
     */
    unregisterTab(tab) {
        this.buttons.update(buttons => buttons.filter(button => button !== tab));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpTabset, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "19.2.11", type: NgpTabset, isStandalone: true, selector: "[ngpTabset]", inputs: { id: { classPropertyName: "id", publicName: "id", isSignal: true, isRequired: false, transformFunction: null }, value: { classPropertyName: "value", publicName: "ngpTabsetValue", isSignal: true, isRequired: false, transformFunction: null }, orientation: { classPropertyName: "orientation", publicName: "ngpTabsetOrientation", isSignal: true, isRequired: false, transformFunction: null }, activateOnFocus: { classPropertyName: "activateOnFocus", publicName: "ngpTabsetActivateOnFocus", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { valueChange: "ngpTabsetValueChange" }, host: { properties: { "attr.id": "state.id()", "attr.data-orientation": "state.orientation()" } }, providers: [provideTabsetState()], exportAs: ["ngpTabset"], hostDirectives: [{ directive: i1.NgpRovingFocusGroup }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpTabset, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpTabset]',
                    exportAs: 'ngpTabset',
                    providers: [provideTabsetState()],
                    hostDirectives: [NgpRovingFocusGroup],
                    host: {
                        '[attr.id]': 'state.id()',
                        '[attr.data-orientation]': 'state.orientation()',
                    },
                }]
        }], ctorParameters: () => [] });

/**
 * Generated bundle index. Do not edit.
 */

export { NgpTabButton, NgpTabList, NgpTabPanel, NgpTabPanelToken, NgpTabset, injectTabPanel, injectTabsetState, provideTabsConfig, provideTabsetState };
