// flowbite-angular/fesm2022/flowbite-angular-indicator.mjs
import { createTheme, mergeDeep, colorToTheme } from 'flowbite-angular';
import * as i0 from '@angular/core';
import { InjectionToken, inject, input, booleanAttribute, computed, Directive } from '@angular/core';
import { createStateToken, createStateProvider, createStateInjector, createState } from 'ng-primitives/state';
import { twMerge } from 'tailwind-merge';

const flowbiteIndicatorTheme = createTheme({
    host: {
        base: 'inline-flex shrink-0 items-center justify-center rounded-full font-bold',
        transition: '',
        border: {
            on: 'border-2',
            off: 'border-none',
        },
        color: {
            default: {
                light: 'bg-gray-500',
                dark: 'dark:bg-gray-500',
            },
            info: {
                light: 'bg-blue-500',
                dark: 'dark:bg-blue-500',
            },
            failure: {
                light: 'bg-red-500',
                dark: 'dark:bg-red-500',
            },
            success: {
                light: 'bg-green-500',
                dark: 'dark:bg-green-500',
            },
            warning: {
                light: 'bg-yellow-500',
                dark: 'dark:bg-yellow-500',
            },
            primary: {
                light: 'bg-primary-500',
                dark: 'dark:bg-primary-500',
            },
        },
        size: {
            xs: 'h-2 w-2 text-xs',
            sm: 'h-3 w-3 text-sm',
            md: 'h-4 w-4 text-sm',
            lg: 'h-5 w-5 text-sm',
            xl: 'h-6 w-6 text-sm',
        },
        position: {
            'top-left': 'top-0 left-0 -translate-x-1/3 -translate-y-1/3',
            'top-center': 'top-0 left-1/2 -translate-x-1/2 -translate-y-1/3',
            'top-right': 'top-0 right-0 translate-x-1/3 -translate-y-1/3',
            'center-left': 'top-1/2 left-0 -translate-x-1/3 -translate-y-1/2',
            center: 'top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2',
            'center-right': 'top-1/2 right-0 translate-x-1/3 -translate-y-1/2',
            'bottom-left': 'bottom-0 left-0 -translate-x-1/3 translate-y-1/3',
            'bottom-center': 'bottom-0 left-1/2 -translate-x-1/2 translate-y-1/3',
            'bottom-right': 'right-0 bottom-0 translate-x-1/3 translate-y-1/3',
        },
    },
});

const defaultFlowbiteIndicatorConfig = {
    baseTheme: flowbiteIndicatorTheme,
    color: 'default',
    size: 'md',
    border: false,
    position: undefined,
    customTheme: {},
};
const FlowbiteIndicatorConfigToken = new InjectionToken('FlowbiteIndicatorConfigToken');
/**
 * Provide the default Indicator configuration
 * @param config The Indicator configuration
 * @returns The provider
 */
const provideFlowbiteIndicatorConfig = (config) => [
    {
        provide: FlowbiteIndicatorConfigToken,
        useValue: { ...defaultFlowbiteIndicatorConfig, ...config },
    },
];
/**
 * Inject the Indicator configuration
 * @see {@link defaultFlowbiteIndicatorConfig}
 * @returns The configuration
 */
const injectFlowbiteIndicatorConfig = () => inject(FlowbiteIndicatorConfigToken, { optional: true }) ?? defaultFlowbiteIndicatorConfig;

const FlowbiteIndicatorStateToken = createStateToken('Flowbite Indicator');
const provideFlowbiteIndicatorState = createStateProvider(FlowbiteIndicatorStateToken);
const injectFlowbiteIndicatorState = createStateInjector(FlowbiteIndicatorStateToken);
const flowbiteIndicatorState = createState(FlowbiteIndicatorStateToken);

class Indicator {
    constructor() {
        this.config = injectFlowbiteIndicatorConfig();
        /**
         * @see {@link injectFlowbiteIndicatorConfig}
         */
        this.color = input(this.config.color);
        /**
         * @see {@link injectFlowbiteIndicatorConfig}
         */
        this.size = input(this.config.size);
        /**
         * @see {@link injectFlowbiteIndicatorConfig}
         */
        this.border = input(this.config.border, { transform: booleanAttribute });
        /**
         * @see {@link injectFlowbiteIndicatorConfig}
         */
        this.position = input(this.config.position);
        /**
         * @see {@link injectFlowbiteIndicatorConfig}
         */
        this.customTheme = input(this.config.customTheme);
        this.theme = computed(() => {
            const mergedTheme = mergeDeep(this.config.baseTheme, this.state.customTheme());
            return {
                host: {
                    root: twMerge(mergedTheme.host.base, colorToTheme(mergedTheme.host.color, this.state.color()), mergedTheme.host.size[this.state.size()], mergedTheme.host.border[this.state.border() ? 'on' : 'off'], this.state.position() && mergedTheme.host.position[this.state.position()]),
                },
            };
        });
        /**
         * @internal
         */
        this.state = flowbiteIndicatorState(this);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: Indicator, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.0.6", type: Indicator, isStandalone: true, selector: "\n    span[flowbiteIndicator]\n  ", inputs: { color: { classPropertyName: "color", publicName: "color", isSignal: true, isRequired: false, transformFunction: null }, size: { classPropertyName: "size", publicName: "size", isSignal: true, isRequired: false, transformFunction: null }, border: { classPropertyName: "border", publicName: "border", isSignal: true, isRequired: false, transformFunction: null }, position: { classPropertyName: "position", publicName: "position", isSignal: true, isRequired: false, transformFunction: null }, customTheme: { classPropertyName: "customTheme", publicName: "customTheme", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "class": "theme().host.root" } }, providers: [provideFlowbiteIndicatorState()], exportAs: ["flowbiteIndicator"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: Indicator, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: `
    span[flowbiteIndicator]
  `,
                    exportAs: 'flowbiteIndicator',
                    hostDirectives: [],
                    providers: [provideFlowbiteIndicatorState()],
                    host: { '[class]': `theme().host.root` },
                }]
        }] });

/* Indicator */

/**
 * Generated bundle index. Do not edit.
 */

export { FlowbiteIndicatorConfigToken, FlowbiteIndicatorStateToken, Indicator, defaultFlowbiteIndicatorConfig, flowbiteIndicatorState, flowbiteIndicatorTheme, injectFlowbiteIndicatorConfig, injectFlowbiteIndicatorState, provideFlowbiteIndicatorConfig, provideFlowbiteIndicatorState };
