// flowbite-angular/fesm2022/flowbite-angular-modal.mjs
import { createTheme, mergeDeep } from 'flowbite-angular';
import * as i0 from '@angular/core';
import { InjectionToken, inject, input, computed, Directive } from '@angular/core';
import { createStateToken, createStateProvider, createStateInjector, createState } from 'ng-primitives/state';
import * as i1 from 'ng-primitives/dialog';
import { NgpDialog, NgpDialogOverlay, NgpDialogTitle, NgpDialogDescription } from 'ng-primitives/dialog';
import { twMerge } from 'tailwind-merge';

const flowbiteModalTheme = createTheme({
    host: {
        base: 'h-max w-full rounded-lg bg-white shadow dark:bg-gray-700',
        transition: '',
        size: {
            sm: 'max-w-md',
            md: 'max-w-lg',
            lg: 'max-w-4xl',
            xl: 'max-w-7xl',
        },
    },
});

const defaultFlowbiteModalConfig = {
    baseTheme: flowbiteModalTheme,
    size: 'md',
    customTheme: {},
};
const FlowbiteModalConfigToken = new InjectionToken('FlowbiteModalConfigToken');
/**
 * Provide the default Modal configuration
 * @param config The Modal configuration
 * @returns The provider
 */
const provideFlowbiteModalConfig = (config) => [
    {
        provide: FlowbiteModalConfigToken,
        useValue: { ...defaultFlowbiteModalConfig, ...config },
    },
];
/**
 * Inject the Modal configuration
 * @see {@link defaultFlowbiteModalConfig}
 * @returns The configuration
 */
const injectFlowbiteModalConfig = () => inject(FlowbiteModalConfigToken, { optional: true }) ?? defaultFlowbiteModalConfig;

const FlowbiteModalStateToken = createStateToken('Flowbite Modal');
const provideFlowbiteModalState = createStateProvider(FlowbiteModalStateToken);
const injectFlowbiteModalState = createStateInjector(FlowbiteModalStateToken);
const flowbiteModalState = createState(FlowbiteModalStateToken);

class Modal {
    constructor() {
        this.config = injectFlowbiteModalConfig();
        /**
         * @see {@link injectFlowbiteModalConfig}
         */
        this.size = input(this.config.size);
        /**
         * @see {@link injectFlowbiteModalConfig}
         */
        this.customTheme = input(this.config.customTheme);
        this.theme = computed(() => {
            const mergedTheme = mergeDeep(this.config.baseTheme, this.state.customTheme());
            return {
                host: {
                    root: twMerge(mergedTheme.host.base, mergedTheme.host.transition, mergedTheme.host.size[this.state.size()]),
                },
            };
        });
        /**
         * @internal
         */
        this.state = flowbiteModalState(this);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: Modal, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.0.6", type: Modal, isStandalone: true, selector: "\n    div[flowbiteModal],\n    section[flowbiteModal]\n  ", inputs: { size: { classPropertyName: "size", publicName: "size", isSignal: true, isRequired: false, transformFunction: null }, customTheme: { classPropertyName: "customTheme", publicName: "customTheme", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "class": "theme().host.root" } }, providers: [provideFlowbiteModalState()], exportAs: ["flowbiteModal"], hostDirectives: [{ directive: i1.NgpDialog, inputs: ["id", "id", "ngpDialogRole", "role", "ngpDialogModal", "modal"] }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: Modal, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: `
    div[flowbiteModal],
    section[flowbiteModal]
  `,
                    exportAs: 'flowbiteModal',
                    hostDirectives: [
                        {
                            directive: NgpDialog,
                            inputs: ['id:id', 'ngpDialogRole:role', 'ngpDialogModal:modal'],
                            outputs: [],
                        },
                    ],
                    providers: [provideFlowbiteModalState()],
                    host: { '[class]': `theme().host.root` },
                }]
        }] });

const flowbiteModalOverlayTheme = createTheme({
    host: {
        base: 'fixed inset-0 flex bg-gray-900/50 p-4 dark:bg-gray-900/80',
        transition: '',
        position: {
            'top-left': 'items-start justify-start',
            'top-center': 'items-start justify-center',
            'top-right': 'items-start justify-end',
            'center-left': 'items-center justify-start',
            center: 'items-center justify-center',
            'center-right': 'items-center justify-end',
            'bottom-left': 'items-end justify-start',
            'bottom-center': 'items-end justify-center',
            'bottom-right': 'items-end justify-end',
        },
    },
});

const defaultFlowbiteModalOverlayConfig = {
    baseTheme: flowbiteModalOverlayTheme,
    position: 'center',
    customTheme: {},
};
const FlowbiteModalOverlayConfigToken = new InjectionToken('FlowbiteModalOverlayConfigToken');
/**
 * Provide the default ModalOverlay configuration
 * @param config The ModalOverlay configuration
 * @returns The provider
 */
const provideFlowbiteModalOverlayConfig = (config) => [
    {
        provide: FlowbiteModalOverlayConfigToken,
        useValue: { ...defaultFlowbiteModalOverlayConfig, ...config },
    },
];
/**
 * Inject the ModalOverlay configuration
 * @see {@link defaultFlowbiteModalOverlayConfig}
 * @returns The configuration
 */
const injectFlowbiteModalOverlayConfig = () => inject(FlowbiteModalOverlayConfigToken, { optional: true }) ?? defaultFlowbiteModalOverlayConfig;

const FlowbiteModalOverlayStateToken = createStateToken('Flowbite ModalOverlay');
const provideFlowbiteModalOverlayState = createStateProvider(FlowbiteModalOverlayStateToken);
const injectFlowbiteModalOverlayState = createStateInjector(FlowbiteModalOverlayStateToken);
const flowbiteModalOverlayState = createState(FlowbiteModalOverlayStateToken);

class ModalOverlay {
    constructor() {
        this.config = injectFlowbiteModalOverlayConfig();
        /**
         * @see {@link injectFlowbiteModalOverlayConfig}
         */
        this.position = input(this.config.position);
        /**
         * @see {@link injectFlowbiteModalOverlayConfig}
         */
        this.customTheme = input(this.config.customTheme);
        this.theme = computed(() => {
            const mergedTheme = mergeDeep(this.config.baseTheme, this.state.customTheme());
            return {
                host: {
                    root: twMerge(mergedTheme.host.base, mergedTheme.host.transition, mergedTheme.host.position[this.state.position()]),
                },
            };
        });
        /**
         * @internal
         */
        this.state = flowbiteModalOverlayState(this);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: ModalOverlay, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.0.6", type: ModalOverlay, isStandalone: true, selector: "\n    div[flowbiteModalOverlay]\n  ", inputs: { position: { classPropertyName: "position", publicName: "position", isSignal: true, isRequired: false, transformFunction: null }, customTheme: { classPropertyName: "customTheme", publicName: "customTheme", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "class": "theme().host.root" } }, providers: [provideFlowbiteModalOverlayState()], exportAs: ["flowbiteModalOverlay"], hostDirectives: [{ directive: i1.NgpDialogOverlay, inputs: ["ngpDialogOverlayCloseOnClick", "closeOnClick"] }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: ModalOverlay, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: `
    div[flowbiteModalOverlay]
  `,
                    exportAs: 'flowbiteModalOverlay',
                    hostDirectives: [
                        {
                            directive: NgpDialogOverlay,
                            inputs: ['ngpDialogOverlayCloseOnClick:closeOnClick'],
                            outputs: [],
                        },
                    ],
                    providers: [provideFlowbiteModalOverlayState()],
                    host: { '[class]': `theme().host.root` },
                }]
        }] });

const flowbiteModalHeaderTheme = createTheme({
    host: {
        base: 'items-center justify-between rounded-t border-b border-gray-200 p-5 text-xl font-medium text-gray-900 dark:border-gray-600 dark:text-white',
        transition: '',
    },
});

const defaultFlowbiteModalHeaderConfig = {
    baseTheme: flowbiteModalHeaderTheme,
    customTheme: {},
};
const FlowbiteModalHeaderConfigToken = new InjectionToken('FlowbiteModalHeaderConfigToken');
/**
 * Provide the default ModalHeader configuration
 * @param config The ModalHeader configuration
 * @returns The provider
 */
const provideFlowbiteModalHeaderConfig = (config) => [
    {
        provide: FlowbiteModalHeaderConfigToken,
        useValue: { ...defaultFlowbiteModalHeaderConfig, ...config },
    },
];
/**
 * Inject the ModalHeader configuration
 * @see {@link defaultFlowbiteModalHeaderConfig}
 * @returns The configuration
 */
const injectFlowbiteModalHeaderConfig = () => inject(FlowbiteModalHeaderConfigToken, { optional: true }) ?? defaultFlowbiteModalHeaderConfig;

const FlowbiteModalHeaderStateToken = createStateToken('Flowbite ModalHeader');
const provideFlowbiteModalHeaderState = createStateProvider(FlowbiteModalHeaderStateToken);
const injectFlowbiteModalHeaderState = createStateInjector(FlowbiteModalHeaderStateToken);
const flowbiteModalHeaderState = createState(FlowbiteModalHeaderStateToken);

class ModalHeader {
    constructor() {
        this.config = injectFlowbiteModalHeaderConfig();
        /**
         * @see {@link injectFlowbiteModalHeaderConfig}
         */
        this.customTheme = input(this.config.customTheme);
        this.theme = computed(() => {
            const mergedTheme = mergeDeep(this.config.baseTheme, this.state.customTheme());
            return {
                host: {
                    root: twMerge(mergedTheme.host.base, mergedTheme.host.transition),
                },
            };
        });
        /**
         * @internal
         */
        this.state = flowbiteModalHeaderState(this);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: ModalHeader, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.0.6", type: ModalHeader, isStandalone: true, selector: "\n    h1[flowbiteModalHeader],\n    h2[flowbiteModalHeader],\n    h3[flowbiteModalHeader],\n    h4[flowbiteModalHeader],\n    h5[flowbiteModalHeader],\n    h6[flowbiteModalHeader]\n  ", inputs: { customTheme: { classPropertyName: "customTheme", publicName: "customTheme", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "class": "theme().host.root" } }, providers: [provideFlowbiteModalHeaderState()], exportAs: ["flowbiteModalHeader"], hostDirectives: [{ directive: i1.NgpDialogTitle, inputs: ["id", "id"] }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: ModalHeader, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: `
    h1[flowbiteModalHeader],
    h2[flowbiteModalHeader],
    h3[flowbiteModalHeader],
    h4[flowbiteModalHeader],
    h5[flowbiteModalHeader],
    h6[flowbiteModalHeader]
  `,
                    exportAs: 'flowbiteModalHeader',
                    hostDirectives: [
                        {
                            directive: NgpDialogTitle,
                            inputs: ['id:id'],
                            outputs: [],
                        },
                    ],
                    providers: [provideFlowbiteModalHeaderState()],
                    host: { '[class]': `theme().host.root` },
                }]
        }] });

const flowbiteModalContentTheme = createTheme({
    host: {
        base: 'block space-y-6 p-6',
        transition: '',
    },
});

const defaultFlowbiteModalContentConfig = {
    baseTheme: flowbiteModalContentTheme,
    customTheme: {},
};
const FlowbiteModalContentConfigToken = new InjectionToken('FlowbiteModalContentConfigToken');
/**
 * Provide the default ModalContent configuration
 * @param config The ModalContent configuration
 * @returns The provider
 */
const provideFlowbiteModalContentConfig = (config) => [
    {
        provide: FlowbiteModalContentConfigToken,
        useValue: { ...defaultFlowbiteModalContentConfig, ...config },
    },
];
/**
 * Inject the ModalContent configuration
 * @see {@link defaultFlowbiteModalContentConfig}
 * @returns The configuration
 */
const injectFlowbiteModalContentConfig = () => inject(FlowbiteModalContentConfigToken, { optional: true }) ?? defaultFlowbiteModalContentConfig;

const FlowbiteModalContentStateToken = createStateToken('Flowbite ModalContent');
const provideFlowbiteModalContentState = createStateProvider(FlowbiteModalContentStateToken);
const injectFlowbiteModalContentState = createStateInjector(FlowbiteModalContentStateToken);
const flowbiteModalContentState = createState(FlowbiteModalContentStateToken);

class ModalContent {
    constructor() {
        this.config = injectFlowbiteModalContentConfig();
        /**
         * @see {@link injectFlowbiteModalContentConfig}
         */
        this.customTheme = input(this.config.customTheme);
        this.theme = computed(() => {
            const mergedTheme = mergeDeep(this.config.baseTheme, this.state.customTheme());
            return {
                host: {
                    root: twMerge(mergedTheme.host.base, mergedTheme.host.transition),
                },
            };
        });
        /**
         * @internal
         */
        this.state = flowbiteModalContentState(this);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: ModalContent, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.0.6", type: ModalContent, isStandalone: true, selector: "\n    [flowbiteModalContent]\n  ", inputs: { customTheme: { classPropertyName: "customTheme", publicName: "customTheme", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "class": "theme().host.root" } }, providers: [provideFlowbiteModalContentState()], exportAs: ["flowbiteModalContent"], hostDirectives: [{ directive: i1.NgpDialogDescription, inputs: ["id", "id"] }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: ModalContent, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: `
    [flowbiteModalContent]
  `,
                    exportAs: 'flowbiteModalContent',
                    hostDirectives: [
                        {
                            directive: NgpDialogDescription,
                            inputs: ['id:id'],
                            outputs: [],
                        },
                    ],
                    providers: [provideFlowbiteModalContentState()],
                    host: { '[class]': `theme().host.root` },
                }]
        }] });

const flowbiteModalFooterTheme = createTheme({
    host: {
        base: 'flex items-center space-x-2 rounded-b border-t border-gray-200 p-6 dark:border-gray-600',
        transition: '',
    },
});

const defaultFlowbiteModalFooterConfig = {
    baseTheme: flowbiteModalFooterTheme,
    customTheme: {},
};
const FlowbiteModalFooterConfigToken = new InjectionToken('FlowbiteModalFooterConfigToken');
/**
 * Provide the default ModalFooter configuration
 * @param config The ModalFooter configuration
 * @returns The provider
 */
const provideFlowbiteModalFooterConfig = (config) => [
    {
        provide: FlowbiteModalFooterConfigToken,
        useValue: { ...defaultFlowbiteModalFooterConfig, ...config },
    },
];
/**
 * Inject the ModalFooter configuration
 * @see {@link defaultFlowbiteModalFooterConfig}
 * @returns The configuration
 */
const injectFlowbiteModalFooterConfig = () => inject(FlowbiteModalFooterConfigToken, { optional: true }) ?? defaultFlowbiteModalFooterConfig;

const FlowbiteModalFooterStateToken = createStateToken('Flowbite ModalFooter');
const provideFlowbiteModalFooterState = createStateProvider(FlowbiteModalFooterStateToken);
const injectFlowbiteModalFooterState = createStateInjector(FlowbiteModalFooterStateToken);
const flowbiteModalFooterState = createState(FlowbiteModalFooterStateToken);

class ModalFooter {
    constructor() {
        this.config = injectFlowbiteModalFooterConfig();
        /**
         * @see {@link injectFlowbiteModalFooterConfig}
         */
        this.customTheme = input(this.config.customTheme);
        this.theme = computed(() => {
            const mergedTheme = mergeDeep(this.config.baseTheme, this.state.customTheme());
            return {
                host: {
                    root: twMerge(mergedTheme.host.base, mergedTheme.host.transition),
                },
            };
        });
        /**
         * @internal
         */
        this.state = flowbiteModalFooterState(this);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: ModalFooter, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.0.6", type: ModalFooter, isStandalone: true, selector: "\n    div[flowbiteModalFooter]\n  ", inputs: { customTheme: { classPropertyName: "customTheme", publicName: "customTheme", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "class": "theme().host.root" } }, providers: [provideFlowbiteModalFooterState()], exportAs: ["flowbiteModalFooter"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: ModalFooter, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: `
    div[flowbiteModalFooter]
  `,
                    exportAs: 'flowbiteModalFooter',
                    hostDirectives: [],
                    providers: [provideFlowbiteModalFooterState()],
                    host: { '[class]': `theme().host.root` },
                }]
        }] });

/* Modal */

/**
 * Generated bundle index. Do not edit.
 */

export { FlowbiteModalConfigToken, FlowbiteModalContentConfigToken, FlowbiteModalContentStateToken, FlowbiteModalFooterConfigToken, FlowbiteModalFooterStateToken, FlowbiteModalHeaderConfigToken, FlowbiteModalHeaderStateToken, FlowbiteModalOverlayConfigToken, FlowbiteModalOverlayStateToken, FlowbiteModalStateToken, Modal, ModalContent, ModalFooter, ModalHeader, ModalOverlay, defaultFlowbiteModalConfig, defaultFlowbiteModalContentConfig, defaultFlowbiteModalFooterConfig, defaultFlowbiteModalHeaderConfig, defaultFlowbiteModalOverlayConfig, flowbiteModalContentState, flowbiteModalContentTheme, flowbiteModalFooterState, flowbiteModalFooterTheme, flowbiteModalHeaderState, flowbiteModalHeaderTheme, flowbiteModalOverlayState, flowbiteModalOverlayTheme, flowbiteModalState, flowbiteModalTheme, injectFlowbiteModalConfig, injectFlowbiteModalContentConfig, injectFlowbiteModalContentState, injectFlowbiteModalFooterConfig, injectFlowbiteModalFooterState, injectFlowbiteModalHeaderConfig, injectFlowbiteModalHeaderState, injectFlowbiteModalOverlayConfig, injectFlowbiteModalOverlayState, injectFlowbiteModalState, provideFlowbiteModalConfig, provideFlowbiteModalContentConfig, provideFlowbiteModalContentState, provideFlowbiteModalFooterConfig, provideFlowbiteModalFooterState, provideFlowbiteModalHeaderConfig, provideFlowbiteModalHeaderState, provideFlowbiteModalOverlayConfig, provideFlowbiteModalOverlayState, provideFlowbiteModalState };
