// flowbite-angular/fesm2022/flowbite-angular-accordion.mjs
import { createTheme, mergeDeep, colorToTheme } from 'flowbite-angular';
import * as i0 from '@angular/core';
import { InjectionToken, inject, input, booleanAttribute, computed, Directive } from '@angular/core';
import { createStateToken, createStateProvider, createStateInjector, createState } from 'ng-primitives/state';
import * as i1 from 'ng-primitives/accordion';
import { provideAccordionState, NgpAccordion, injectAccordionItemState, provideAccordionItemState, NgpAccordionItem, NgpAccordionTrigger, NgpAccordionContent } from 'ng-primitives/accordion';
import { twMerge } from 'tailwind-merge';
import * as i1$1 from 'ng-primitives/button';
import { NgpButton } from 'ng-primitives/button';
import * as i2 from 'ng-primitives/interactions';
import { NgpFocus } from 'ng-primitives/interactions';

const flowbiteAccordionTheme = createTheme({
    host: {
        base: 'shadow-sm',
        transition: '',
    },
});

const defaultFlowbiteAccordionConfig = {
    baseTheme: flowbiteAccordionTheme,
    color: 'default',
    flush: false,
    customTheme: {},
};
const FlowbiteAccordionConfigToken = new InjectionToken('FlowbiteAccordionConfig');
/**
 * Provide the default Accordion configuration
 * @param config The Accordion configuration
 * @returns The provider
 */
const provideFlowbiteAccordionConfig = (config) => [
    {
        provide: FlowbiteAccordionConfigToken,
        useValue: { ...defaultFlowbiteAccordionConfig, ...config },
    },
];
/**
 * Inject the Accordion configuration
 * @see {@link defaultFlowbiteAccordionConfig}
 * @returns The configuration
 */
const injectFlowbiteAccordionConfig = () => inject(FlowbiteAccordionConfigToken, { optional: true }) ?? defaultFlowbiteAccordionConfig;

const FlowbiteAccordionStateToken = createStateToken('Flowbite Accordion');
const provideFlowbiteAccordionState = createStateProvider(FlowbiteAccordionStateToken);
const injectFlowbiteAccordionState = createStateInjector(FlowbiteAccordionStateToken);
const flowbiteAccordionState = createState(FlowbiteAccordionStateToken);

class Accordion {
    constructor() {
        this.config = injectFlowbiteAccordionConfig();
        /**
         * @see {@link injectFlowbiteAccordionConfig}
         */
        this.color = input(this.config.color);
        /**
         * @see {@link injectFlowbiteAccordionConfig}
         */
        this.flush = input(this.config.flush, { transform: booleanAttribute });
        /**
         * @see {@link injectFlowbiteAccordionConfig}
         */
        this.customTheme = input(this.config.customTheme);
        this.theme = computed(() => {
            const mergedTheme = mergeDeep(this.config.baseTheme, this.state.customTheme());
            return {
                host: {
                    root: twMerge(mergedTheme.host.base, mergedTheme.host.transition),
                },
            };
        });
        /**
         * @internal
         */
        this.state = flowbiteAccordionState(this);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: Accordion, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.0.6", type: Accordion, isStandalone: true, selector: "\n    div[flowbiteAccordion]\n  ", inputs: { color: { classPropertyName: "color", publicName: "color", isSignal: true, isRequired: false, transformFunction: null }, flush: { classPropertyName: "flush", publicName: "flush", isSignal: true, isRequired: false, transformFunction: null }, customTheme: { classPropertyName: "customTheme", publicName: "customTheme", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "class": "theme().host.root" } }, providers: [provideFlowbiteAccordionState(), provideAccordionState()], exportAs: ["flowbiteAccordion"], hostDirectives: [{ directive: i1.NgpAccordion, inputs: ["ngpAccordionType", "type", "ngpAccordionDisabled", "disabled", "ngpAccordionCollapsible", "collapsible"] }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: Accordion, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: `
    div[flowbiteAccordion]
  `,
                    exportAs: 'flowbiteAccordion',
                    hostDirectives: [
                        {
                            directive: NgpAccordion,
                            inputs: [
                                'ngpAccordionType:type',
                                'ngpAccordionDisabled:disabled',
                                'ngpAccordionCollapsible:collapsible',
                            ],
                        },
                    ],
                    providers: [provideFlowbiteAccordionState(), provideAccordionState()],
                    host: { '[class]': `theme().host.root` },
                }]
        }] });

const flowbiteAccordionItemTheme = createTheme({
    host: {
        base: 'group/item',
        transition: '',
    },
});

const defaultFlowbiteAccordionItemConfig = {
    baseTheme: flowbiteAccordionItemTheme,
    customTheme: {},
};
const FlowbiteAccordionItemConfigToken = new InjectionToken('FlowbiteAccordionItemConfig');
/**
 * Provide the default Accordion Item configuration
 * @param config The Accordion Item configuration
 * @returns The provider
 */
const provideFlowbiteAccordionItemConfig = (config) => [
    {
        provide: FlowbiteAccordionItemConfigToken,
        useValue: { ...defaultFlowbiteAccordionItemConfig, ...config },
    },
];
/**
 * Inject the Accordion Item configuration
 * @see {@link defaultFlowbiteAccordionItemConfig}
 * @returns The configuration
 */
const injectFlowbiteAccordionItemConfig = () => inject(FlowbiteAccordionItemConfigToken, { optional: true }) ??
    defaultFlowbiteAccordionItemConfig;

const FlowbiteAccordionItemStateToken = createStateToken('Flowbite Accordion Item');
const provideFlowbiteAccordionItemState = createStateProvider(FlowbiteAccordionItemStateToken);
const injectFlowbiteAccordionItemState = createStateInjector(FlowbiteAccordionItemStateToken);
const flowbiteAccordionItemState = createState(FlowbiteAccordionItemStateToken);

class AccordionItem {
    constructor() {
        this.config = injectFlowbiteAccordionItemConfig();
        this.accordionState = injectFlowbiteAccordionState();
        this.ngpAccordionItemState = injectAccordionItemState();
        /**
         * @see {@link injectFlowbiteAccordionItemConfig}
         */
        this.customTheme = input(this.config.customTheme);
        this.theme = computed(() => {
            const mergedTheme = mergeDeep(this.config.baseTheme, this.state.customTheme());
            return {
                host: {
                    root: twMerge(mergedTheme.host.base, mergedTheme.host.transition),
                },
            };
        });
        /**
         * @internal
         */
        this.state = flowbiteAccordionItemState(this);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: AccordionItem, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.0.6", type: AccordionItem, isStandalone: true, selector: "\n    div[flowbiteAccordionItem]\n  ", inputs: { customTheme: { classPropertyName: "customTheme", publicName: "customTheme", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "class": "theme().host.root" } }, providers: [provideFlowbiteAccordionItemState(), provideAccordionItemState()], exportAs: ["flowbiteAccordionItem"], hostDirectives: [{ directive: i1.NgpAccordionItem, inputs: ["ngpAccordionItemValue", "value", "ngpAccordionItemDisabled", "disabled"] }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: AccordionItem, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: `
    div[flowbiteAccordionItem]
  `,
                    exportAs: 'flowbiteAccordionItem',
                    hostDirectives: [
                        {
                            directive: NgpAccordionItem,
                            inputs: ['ngpAccordionItemValue:value', 'ngpAccordionItemDisabled:disabled'],
                        },
                    ],
                    providers: [provideFlowbiteAccordionItemState(), provideAccordionItemState()],
                    host: { '[class]': `theme().host.root` },
                }]
        }] });

const flowbiteAccordionTitleTheme = createTheme({
    host: {
        base: 'flex w-full cursor-pointer flex-row items-center justify-between p-5 text-base font-medium group-first/item:rounded-t-lg',
        transition: '',
        color: {
            default: {
                light: 'border-gray-200 bg-gray-100 text-gray-900 data-hover:bg-gray-200',
                dark: 'dark:border-gray-700 dark:bg-gray-800 dark:text-white dark:data-hover:bg-gray-700',
            },
            info: {
                light: 'border-blue-200 bg-blue-100 text-blue-900 data-hover:bg-blue-200',
                dark: 'dark:border-blue-700 dark:bg-blue-800 dark:text-white dark:data-hover:bg-blue-700',
            },
            failure: {
                light: 'border-red-200 bg-red-100 text-red-900 data-hover:bg-red-200',
                dark: 'dark:border-red-700 dark:bg-red-800 dark:text-white dark:data-hover:bg-red-700',
            },
            success: {
                light: 'border-green-200 bg-green-100 text-green-900 data-hover:bg-green-200',
                dark: 'dark:border-green-700 dark:bg-green-800 dark:text-white dark:data-hover:bg-green-700',
            },
            warning: {
                light: 'border-yellow-200 bg-yellow-100 text-yellow-900 data-hover:bg-yellow-200',
                dark: 'dark:border-yellow-700 dark:bg-yellow-800 dark:text-white dark:data-hover:bg-yellow-700',
            },
            primary: {
                light: 'border-primary-200 bg-primary-100 text-primary-900 data-hover:bg-primary-200',
                dark: 'dark:border-primary-700 dark:bg-primary-800 dark:data-hover:bg-primary-700 dark:text-white',
            },
        },
        flush: {
            on: 'border-b bg-transparent data-hover:bg-transparent dark:bg-transparent dark:data-hover:bg-transparent',
            off: 'border',
        },
    },
});

const defaultFlowbiteAccordionTitleConfig = {
    baseTheme: flowbiteAccordionTitleTheme,
    customTheme: {},
};
const FlowbiteAccordionTitleConfigToken = new InjectionToken('FlowbiteAccordionTitleConfigToken');
/**
 * Provide the default AccordionTitle configuration
 * @param config The AccordionTitle configuration
 * @returns The provider
 */
const provideFlowbiteAccordionTitleConfig = (config) => [
    {
        provide: FlowbiteAccordionTitleConfigToken,
        useValue: { ...defaultFlowbiteAccordionTitleConfig, ...config },
    },
];
/**
 * Inject the AccordionTitle configuration
 * @see {@link defaultFlowbiteAccordionTitleConfig}
 * @returns The configuration
 */
const injectFlowbiteAccordionTitleConfig = () => inject(FlowbiteAccordionTitleConfigToken, { optional: true }) ??
    defaultFlowbiteAccordionTitleConfig;

const FlowbiteAccordionTitleStateToken = createStateToken('Flowbite AccordionTitle');
const provideFlowbiteAccordionTitleState = createStateProvider(FlowbiteAccordionTitleStateToken);
const injectFlowbiteAccordionTitleState = createStateInjector(FlowbiteAccordionTitleStateToken);
const flowbiteAccordionTitleState = createState(FlowbiteAccordionTitleStateToken);

class AccordionTitle {
    constructor() {
        this.config = injectFlowbiteAccordionTitleConfig();
        this.accordionState = injectFlowbiteAccordionState();
        /**
         * @see {@link injectFlowbiteAccordionTitleConfig}
         */
        this.customTheme = input(this.config.customTheme);
        this.theme = computed(() => {
            const mergedTheme = mergeDeep(this.config.baseTheme, this.state.customTheme());
            return {
                host: {
                    root: twMerge(mergedTheme.host.base, mergedTheme.host.transition, colorToTheme(mergedTheme.host.color, this.accordionState().color()), mergedTheme.host.flush[this.accordionState().flush() ? 'on' : 'off']),
                },
            };
        });
        /**
         * @internal
         */
        this.state = flowbiteAccordionTitleState(this);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: AccordionTitle, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.0.6", type: AccordionTitle, isStandalone: true, selector: "\n    [flowbiteAccordionTitle]\n  ", inputs: { customTheme: { classPropertyName: "customTheme", publicName: "customTheme", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "class": "theme().host.root" } }, providers: [provideFlowbiteAccordionTitleState()], exportAs: ["flowbiteAccordionTitle"], hostDirectives: [{ directive: i1$1.NgpButton, inputs: ["disabled", "disabled"] }, { directive: i2.NgpFocus }, { directive: i1.NgpAccordionTrigger, inputs: ["id", "id"] }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: AccordionTitle, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: `
    [flowbiteAccordionTitle]
  `,
                    exportAs: 'flowbiteAccordionTitle',
                    hostDirectives: [
                        {
                            directive: NgpButton,
                            inputs: ['disabled:disabled'],
                            outputs: [],
                        },
                        {
                            directive: NgpFocus,
                            inputs: [],
                            outputs: [],
                        },
                        {
                            directive: NgpAccordionTrigger,
                            inputs: ['id'],
                            outputs: [],
                        },
                    ],
                    providers: [provideFlowbiteAccordionTitleState()],
                    host: { '[class]': `theme().host.root` },
                }]
        }] });

const flowbiteAccordionContentTheme = createTheme({
    host: {
        base: 'overflow-hidden',
        transition: '',
        color: {
            default: {
                light: 'border-gray-200 bg-white',
                dark: 'dark:border-gray-700 dark:bg-gray-900',
            },
            info: {
                light: 'border-blue-200 bg-white',
                dark: 'dark:border-blue-700 dark:bg-gray-900',
            },
            failure: {
                light: 'border-red-200 bg-white',
                dark: 'dark:border-red-700 dark:bg-gray-900',
            },
            success: {
                light: 'border-green-200 bg-white',
                dark: 'dark:border-green-700 dark:bg-gray-900',
            },
            warning: {
                light: 'border-yellow-200 bg-white',
                dark: 'dark:border-yellow-700 dark:bg-gray-900',
            },
            primary: {
                light: 'border-primary-200 bg-white',
                dark: 'dark:border-primary-700 dark:bg-gray-900',
            },
        },
        open: {
            on: 'h-fit',
            off: 'h-0',
        },
        flush: {
            on: 'bg-transparent dark:bg-transparent',
            off: 'border-x group-last/item:border-b',
        },
    },
});

const defaultFlowbiteAccordionContentConfig = {
    baseTheme: flowbiteAccordionContentTheme,
    customTheme: {},
};
const FlowbiteAccordionContentConfigToken = new InjectionToken('FlowbiteAccordionContentConfigToken');
/**
 * Provide the default AccordionContent configuration
 * @param config The AccordionContent configuration
 * @returns The provider
 */
const provideFlowbiteAccordionContentConfig = (config) => [
    {
        provide: FlowbiteAccordionContentConfigToken,
        useValue: { ...defaultFlowbiteAccordionContentConfig, ...config },
    },
];
/**
 * Inject the AccordionContent configuration
 * @see {@link defaultFlowbiteAccordionContentConfig}
 * @returns The configuration
 */
const injectFlowbiteAccordionContentConfig = () => inject(FlowbiteAccordionContentConfigToken, { optional: true }) ??
    defaultFlowbiteAccordionContentConfig;

const FlowbiteAccordionContentStateToken = createStateToken('Flowbite AccordionContent');
const provideFlowbiteAccordionContentState = createStateProvider(FlowbiteAccordionContentStateToken);
const injectFlowbiteAccordionContentState = createStateInjector(FlowbiteAccordionContentStateToken);
const flowbiteAccordionContentState = createState(FlowbiteAccordionContentStateToken);

class AccordionContent {
    constructor() {
        this.config = injectFlowbiteAccordionContentConfig();
        this.accordionState = injectFlowbiteAccordionState();
        this.ngpAccordionItemState = injectAccordionItemState();
        /**
         * @see {@link injectFlowbiteAccordionContentConfig}
         */
        this.customTheme = input(this.config.customTheme);
        this.theme = computed(() => {
            const mergedTheme = mergeDeep(this.config.baseTheme, this.state.customTheme());
            return {
                host: {
                    root: twMerge(mergedTheme.host.base, mergedTheme.host.transition, colorToTheme(mergedTheme.host.color, this.accordionState().color()), mergedTheme.host.flush[this.accordionState().flush() ? 'on' : 'off'], mergedTheme.host.open[this.ngpAccordionItemState().open() ? 'on' : 'off']),
                },
            };
        });
        /**
         * @internal
         */
        this.state = flowbiteAccordionContentState(this);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: AccordionContent, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.0.6", type: AccordionContent, isStandalone: true, selector: "\n    div[flowbiteAccordionContent]\n  ", inputs: { customTheme: { classPropertyName: "customTheme", publicName: "customTheme", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "class": "theme().host.root" } }, providers: [provideFlowbiteAccordionContentState()], exportAs: ["flowbiteAccordionContent"], hostDirectives: [{ directive: i1.NgpAccordionContent, inputs: ["id", "id"] }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: AccordionContent, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: `
    div[flowbiteAccordionContent]
  `,
                    exportAs: 'flowbiteAccordionContent',
                    hostDirectives: [
                        {
                            directive: NgpAccordionContent,
                            inputs: ['id'],
                            outputs: [],
                        },
                    ],
                    providers: [provideFlowbiteAccordionContentState()],
                    host: { '[class]': `theme().host.root` },
                }]
        }] });

/* Accordion */

/**
 * Generated bundle index. Do not edit.
 */

export { Accordion, AccordionContent, AccordionItem, AccordionTitle, FlowbiteAccordionConfigToken, FlowbiteAccordionContentConfigToken, FlowbiteAccordionContentStateToken, FlowbiteAccordionItemConfigToken, FlowbiteAccordionItemStateToken, FlowbiteAccordionStateToken, FlowbiteAccordionTitleConfigToken, FlowbiteAccordionTitleStateToken, defaultFlowbiteAccordionConfig, defaultFlowbiteAccordionContentConfig, defaultFlowbiteAccordionItemConfig, defaultFlowbiteAccordionTitleConfig, flowbiteAccordionContentState, flowbiteAccordionContentTheme, flowbiteAccordionItemState, flowbiteAccordionItemTheme, flowbiteAccordionState, flowbiteAccordionTheme, flowbiteAccordionTitleState, flowbiteAccordionTitleTheme, injectFlowbiteAccordionConfig, injectFlowbiteAccordionContentConfig, injectFlowbiteAccordionContentState, injectFlowbiteAccordionItemConfig, injectFlowbiteAccordionItemState, injectFlowbiteAccordionState, injectFlowbiteAccordionTitleConfig, injectFlowbiteAccordionTitleState, provideFlowbiteAccordionConfig, provideFlowbiteAccordionContentConfig, provideFlowbiteAccordionContentState, provideFlowbiteAccordionItemConfig, provideFlowbiteAccordionItemState, provideFlowbiteAccordionState, provideFlowbiteAccordionTitleConfig, provideFlowbiteAccordionTitleState };
