// flowbite-angular/fesm2022/flowbite-angular-sidebar.mjs
import { createTheme, mergeDeep, colorToTheme } from 'flowbite-angular';
import * as i0 from '@angular/core';
import { InjectionToken, inject, input, booleanAttribute, computed, Directive, ChangeDetectionStrategy, ViewEncapsulation, Component } from '@angular/core';
import { createStateToken, createStateProvider, createStateInjector, createState } from 'ng-primitives/state';
import { twMerge } from 'tailwind-merge';
import * as i1 from 'flowbite-angular/button';
import { BaseButton } from 'flowbite-angular/button';
import { barsFromLeft } from 'flowbite-angular/icon/outline/general';
import { provideIcons } from '@ng-icons/core';
import * as i2 from 'ng-primitives/button';
import { NgpButton } from 'ng-primitives/button';
import * as i3 from 'ng-primitives/interactions';
import { NgpFocus } from 'ng-primitives/interactions';

const flowbiteSidebarTheme = createTheme({
    host: {
        base: 'fixed inset-0 top-0 left-0 z-40 h-screen lg:w-64',
        transition: '',
        open: {
            on: 'w-full backdrop-blur-sm',
            off: 'w-0',
        },
    },
});

const defaultFlowbiteSidebarConfig = {
    baseTheme: flowbiteSidebarTheme,
    open: false,
    color: 'default',
    customTheme: {},
};
const FlowbiteSidebarConfigToken = new InjectionToken('FlowbiteSidebarConfigToken');
/**
 * Provide the default Sidebar configuration
 * @param config The Sidebar configuration
 * @returns The provider
 */
const provideFlowbiteSidebarConfig = (config) => [
    {
        provide: FlowbiteSidebarConfigToken,
        useValue: { ...defaultFlowbiteSidebarConfig, ...config },
    },
];
/**
 * Inject the Sidebar configuration
 * @see {@link defaultFlowbiteSidebarConfig}
 * @returns The configuration
 */
const injectFlowbiteSidebarConfig = () => inject(FlowbiteSidebarConfigToken, { optional: true }) ?? defaultFlowbiteSidebarConfig;

const FlowbiteSidebarStateToken = createStateToken('Flowbite Sidebar');
const provideFlowbiteSidebarState = createStateProvider(FlowbiteSidebarStateToken);
const injectFlowbiteSidebarState = createStateInjector(FlowbiteSidebarStateToken);
const flowbiteSidebarState = createState(FlowbiteSidebarStateToken);

class Sidebar {
    constructor() {
        this.config = injectFlowbiteSidebarConfig();
        /**
         * @see {@link injectFlowbiteSidebarConfig}
         */
        this.open = input(this.config.open, { transform: booleanAttribute });
        /**
         * @see {@link injectFlowbiteSidebarConfig}
         */
        this.color = input(this.config.color);
        /**
         * @see {@link injectFlowbiteSidebarConfig}
         */
        this.customTheme = input(this.config.customTheme);
        this.theme = computed(() => {
            const mergedTheme = mergeDeep(this.config.baseTheme, this.state.customTheme());
            return {
                host: {
                    root: twMerge(mergedTheme.host.base, mergedTheme.host.transition, mergedTheme.host.open[this.state.open() ? 'on' : 'off']),
                },
            };
        });
        /**
         * @internal
         */
        this.state = flowbiteSidebarState(this);
    }
    /**
     * @internal
     */
    onClick() {
        this.toggle();
    }
    /**
     * @internal
     */
    toggle(newState) {
        newState ??= !this.state.open();
        this.state.open.set(newState);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: Sidebar, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.0.6", type: Sidebar, isStandalone: true, selector: "\n    aside[flowbiteSidebar]\n  ", inputs: { open: { classPropertyName: "open", publicName: "open", isSignal: true, isRequired: false, transformFunction: null }, color: { classPropertyName: "color", publicName: "color", isSignal: true, isRequired: false, transformFunction: null }, customTheme: { classPropertyName: "customTheme", publicName: "customTheme", isSignal: true, isRequired: false, transformFunction: null } }, host: { listeners: { "click": "onClick()" }, properties: { "class": "theme().host.root" } }, providers: [provideFlowbiteSidebarState({ inherit: true })], exportAs: ["flowbiteSidebar"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: Sidebar, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: `
    aside[flowbiteSidebar]
  `,
                    exportAs: 'flowbiteSidebar',
                    hostDirectives: [],
                    providers: [provideFlowbiteSidebarState({ inherit: true })],
                    host: {
                        '[class]': `theme().host.root`,
                        '(click)': 'onClick()',
                    },
                }]
        }] });

const flowbiteSidebarContentTheme = createTheme({
    host: {
        base: 'h-full overflow-y-auto border-r px-3 py-4 lg:translate-x-0 lg:border-0',
        transition: '',
        open: {
            on: 'w-64 translate-x-0',
            off: '-translate-x-full',
        },
        color: {
            default: {
                light: 'border-gray-200 bg-gray-100',
                dark: 'dark:border-gray-800 dark:bg-gray-900',
            },
            info: {
                light: 'border-blue-200 bg-gray-100',
                dark: 'dark:border-blue-800 dark:bg-gray-900',
            },
            failure: {
                light: 'border-red-200 bg-gray-100',
                dark: 'dark:border-red-800 dark:bg-gray-900',
            },
            success: {
                light: 'border-green-200 bg-gray-100',
                dark: 'dark:border-green-800 dark:bg-gray-900',
            },
            warning: {
                light: 'border-yellow-200 bg-gray-100',
                dark: 'dark:border-yellow-800 dark:bg-gray-900',
            },
            primary: {
                light: 'border-primary-200 bg-gray-100',
                dark: 'dark:border-primary-800 dark:bg-gray-900',
            },
        },
    },
    container: {
        base: 'space-y-2 font-medium',
    },
});

const defaultFlowbiteSidebarContentConfig = {
    baseTheme: flowbiteSidebarContentTheme,
    customTheme: {},
};
const FlowbiteSidebarContentConfigToken = new InjectionToken('FlowbiteSidebarContentConfigToken');
/**
 * Provide the default SidebarContent configuration
 * @param config The SidebarContent configuration
 * @returns The provider
 */
const provideFlowbiteSidebarContentConfig = (config) => [
    {
        provide: FlowbiteSidebarContentConfigToken,
        useValue: { ...defaultFlowbiteSidebarContentConfig, ...config },
    },
];
/**
 * Inject the SidebarContent configuration
 * @see {@link defaultFlowbiteSidebarContentConfig}
 * @returns The configuration
 */
const injectFlowbiteSidebarContentConfig = () => inject(FlowbiteSidebarContentConfigToken, { optional: true }) ??
    defaultFlowbiteSidebarContentConfig;

const FlowbiteSidebarContentStateToken = createStateToken('Flowbite SidebarContent');
const provideFlowbiteSidebarContentState = createStateProvider(FlowbiteSidebarContentStateToken);
const injectFlowbiteSidebarContentState = createStateInjector(FlowbiteSidebarContentStateToken);
const flowbiteSidebarContentState = createState(FlowbiteSidebarContentStateToken);

class SidebarContent {
    constructor() {
        this.config = injectFlowbiteSidebarContentConfig();
        this.sidebarState = injectFlowbiteSidebarState();
        /**
         * @see {@link injectFlowbiteSidebarContentConfig}
         */
        this.customTheme = input(this.config.customTheme);
        this.theme = computed(() => {
            const mergedTheme = mergeDeep(this.config.baseTheme, this.state.customTheme());
            return {
                host: {
                    root: twMerge(mergedTheme.host.base, mergedTheme.host.transition, mergedTheme.host.open[this.sidebarState().open() ? 'on' : 'off'], colorToTheme(mergedTheme.host.color, this.sidebarState().color())),
                },
                container: {
                    root: twMerge(mergedTheme.container.base),
                },
            };
        });
        /**
         * @internal
         */
        this.state = flowbiteSidebarContentState(this);
    }
    /**
     * @internal
     */
    onClick($event) {
        $event.stopPropagation();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: SidebarContent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.1.0", version: "20.0.6", type: SidebarContent, isStandalone: true, selector: "\n    div[flowbiteSidebarContent]\n  ", inputs: { customTheme: { classPropertyName: "customTheme", publicName: "customTheme", isSignal: true, isRequired: false, transformFunction: null } }, host: { listeners: { "click": "onClick($event)" }, properties: { "class": "theme().host.root" } }, providers: [provideFlowbiteSidebarContentState()], exportAs: ["flowbiteSidebarContent"], ngImport: i0, template: `
    <ul [class]="theme().container.root">
      <ng-content />
    </ul>
  `, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: SidebarContent, decorators: [{
            type: Component,
            args: [{
                    standalone: true,
                    selector: `
    div[flowbiteSidebarContent]
  `,
                    exportAs: 'flowbiteSidebarContent',
                    hostDirectives: [],
                    imports: [],
                    providers: [provideFlowbiteSidebarContentState()],
                    host: {
                        '[class]': `theme().host.root`,
                        '(click)': 'onClick($event)',
                    },
                    template: `
    <ul [class]="theme().container.root">
      <ng-content />
    </ul>
  `,
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }] });

const flowbiteSidebarToggleTheme = createTheme({
    host: {
        base: 'inline-flex h-10 w-10 cursor-pointer items-center justify-center rounded-lg p-2 text-sm lg:hidden',
        transition: '',
        focus: 'data-focus:ring-0 data-focus:outline-none data-focus-visible:ring-2 data-focus-visible:outline-none',
        disabled: 'data-disabled:cursor-not-allowed data-disabled:opacity-50',
        size: {},
        pill: {},
        color: {
            default: {
                light: 'text-gray-900',
                dark: 'dark:text-gray-100',
            },
            info: {
                light: 'text-blue-900',
                dark: 'dark:text-blue-100',
            },
            failure: {
                light: 'text-red-900',
                dark: 'dark:text-red-100',
            },
            success: {
                light: 'text-green-900',
                dark: 'dark:text-green-100',
            },
            warning: {
                light: 'text-yellow-900',
                dark: 'dark:text-yellow-100',
            },
            primary: {
                light: 'text-primary-900',
                dark: 'dark:text-primary-100',
            },
        },
        colorOutline: {},
    },
});

const defaultFlowbiteSidebarToggleConfig = {
    baseTheme: flowbiteSidebarToggleTheme,
    customTheme: {},
};
const FlowbiteSidebarToggleConfigToken = new InjectionToken('FlowbiteSidebarToggleConfigToken');
/**
 * Provide the default SidebarToggle configuration
 * @param config The SidebarToggle configuration
 * @returns The provider
 */
const provideFlowbiteSidebarToggleConfig = (config) => [
    {
        provide: FlowbiteSidebarToggleConfigToken,
        useValue: { ...defaultFlowbiteSidebarToggleConfig, ...config },
    },
];
/**
 * Inject the SidebarToggle configuration
 * @see {@link defaultFlowbiteSidebarToggleConfig}
 * @returns The configuration
 */
const injectFlowbiteSidebarToggleConfig = () => inject(FlowbiteSidebarToggleConfigToken, { optional: true }) ??
    defaultFlowbiteSidebarToggleConfig;

const FlowbiteSidebarToggleStateToken = createStateToken('Flowbite SidebarToggle');
const provideFlowbiteSidebarToggleState = createStateProvider(FlowbiteSidebarToggleStateToken);
const injectFlowbiteSidebarToggleState = createStateInjector(FlowbiteSidebarToggleStateToken);
const flowbiteSidebarToggleState = createState(FlowbiteSidebarToggleStateToken);

class SidebarToggle {
    constructor() {
        this.config = injectFlowbiteSidebarToggleConfig();
        this.sidebarState = injectFlowbiteSidebarState();
        /**
         * @see {@link injectFlowbiteSidebarToggleConfig}
         */
        this.customTheme = input(this.config.customTheme);
        this.theme = computed(() => {
            const mergedTheme = mergeDeep(this.config.baseTheme, this.state.customTheme());
            return {
                host: {
                    root: twMerge(mergedTheme.host.base, mergedTheme.host.transition, mergedTheme.host.focus, mergedTheme.host.disabled, colorToTheme(mergedTheme.host.color, this.sidebarState().color())),
                },
            };
        });
        /**
         * @internal
         */
        this.state = flowbiteSidebarToggleState(this);
    }
    /**
     * @internal
     */
    onClick() {
        this.toggleSidebar();
    }
    /**
     * @internal
     */
    toggleSidebar() {
        this.sidebarState().toggle();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: SidebarToggle, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.0.6", type: SidebarToggle, isStandalone: true, selector: "\n    button[flowbiteSidebarToggle]\n  ", inputs: { customTheme: { classPropertyName: "customTheme", publicName: "customTheme", isSignal: true, isRequired: false, transformFunction: null } }, host: { listeners: { "click": "onClick()" }, properties: { "class": "theme().host.root" } }, providers: [provideFlowbiteSidebarToggleState(), provideIcons({ barsFromLeft })], exportAs: ["flowbiteSidebarToggle"], hostDirectives: [{ directive: i1.BaseButton }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: SidebarToggle, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: `
    button[flowbiteSidebarToggle]
  `,
                    exportAs: 'flowbiteSidebarToggle',
                    hostDirectives: [
                        {
                            directive: BaseButton,
                            inputs: [],
                            outputs: [],
                        },
                    ],
                    providers: [provideFlowbiteSidebarToggleState(), provideIcons({ barsFromLeft })],
                    host: {
                        '[class]': `theme().host.root`,
                        '(click)': 'onClick()',
                    },
                }]
        }] });

const flowbiteSidebarItemTheme = createTheme({
    host: {
        base: 'm-0 flex cursor-pointer flex-row rounded-sm px-3 py-2',
        transition: '',
        focus: 'data-focus:ring-0 data-focus:outline-none data-focus-visible:ring-2 data-focus-visible:outline-none',
        disabled: 'data-disabled:cursor-not-allowed data-disabled:opacity-50',
        size: {},
        pill: {},
        color: {
            default: {
                light: 'text-gray-800 data-hover:text-gray-900',
                dark: 'dark:text-white dark:data-hover:text-gray-100',
            },
            info: {
                light: 'text-gray-800 data-hover:text-blue-900',
                dark: 'dark:text-white dark:data-hover:text-blue-100',
            },
            failure: {
                light: 'text-gray-800 data-hover:text-red-900',
                dark: 'dark:text-white dark:data-hover:text-red-100',
            },
            success: {
                light: 'text-gray-800 data-hover:text-green-900',
                dark: 'dark:text-white dark:data-hover:text-green-100',
            },
            warning: {
                light: 'text-gray-800 data-hover:text-yellow-900',
                dark: 'dark:text-white dark:data-hover:text-yellow-100',
            },
            primary: {
                light: 'data-hover:text-primary-900 text-gray-800',
                dark: 'dark:data-hover:text-primary-100 dark:text-white',
            },
        },
        colorOutline: {},
    },
});

const defaultFlowbiteSidebarItemConfig = {
    baseTheme: flowbiteSidebarItemTheme,
    customTheme: {},
};
const FlowbiteSidebarItemConfigToken = new InjectionToken('FlowbiteSidebarItemConfigToken');
/**
 * Provide the default SidebarItem configuration
 * @param config The SidebarItem configuration
 * @returns The provider
 */
const provideFlowbiteSidebarItemConfig = (config) => [
    {
        provide: FlowbiteSidebarItemConfigToken,
        useValue: { ...defaultFlowbiteSidebarItemConfig, ...config },
    },
];
/**
 * Inject the SidebarItem configuration
 * @see {@link defaultFlowbiteSidebarItemConfig}
 * @returns The configuration
 */
const injectFlowbiteSidebarItemConfig = () => inject(FlowbiteSidebarItemConfigToken, { optional: true }) ?? defaultFlowbiteSidebarItemConfig;

const FlowbiteSidebarItemStateToken = createStateToken('Flowbite SidebarItem');
const provideFlowbiteSidebarItemState = createStateProvider(FlowbiteSidebarItemStateToken);
const injectFlowbiteSidebarItemState = createStateInjector(FlowbiteSidebarItemStateToken);
const flowbiteSidebarItemState = createState(FlowbiteSidebarItemStateToken);

class SidebarItem {
    constructor() {
        this.config = injectFlowbiteSidebarItemConfig();
        this.sidebarState = injectFlowbiteSidebarState();
        /**
         * @see {@link injectFlowbiteSidebarItemConfig}
         */
        this.customTheme = input(this.config.customTheme);
        this.theme = computed(() => {
            const mergedTheme = mergeDeep(this.config.baseTheme, this.state.customTheme());
            return {
                host: {
                    root: twMerge(mergedTheme.host.base, mergedTheme.host.transition, mergedTheme.host.focus, mergedTheme.host.disabled, colorToTheme(mergedTheme.host.color, this.sidebarState().color())),
                },
            };
        });
        /**
         * @internal
         */
        this.state = flowbiteSidebarItemState(this);
    }
    /**
     * @internal
     */
    onClick() {
        this.toggleSidebar();
    }
    /**
     * @internal
     */
    toggleSidebar() {
        this.sidebarState().toggle();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: SidebarItem, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.0.6", type: SidebarItem, isStandalone: true, selector: "\n    a[flowbiteSidebarItem],\n    button[flowbiteSidebarItem]\n  ", inputs: { customTheme: { classPropertyName: "customTheme", publicName: "customTheme", isSignal: true, isRequired: false, transformFunction: null } }, host: { listeners: { "click": "onClick()" }, properties: { "class": "theme().host.root" } }, providers: [provideFlowbiteSidebarItemState()], exportAs: ["flowbiteSidebarItem"], hostDirectives: [{ directive: i1.BaseButton }, { directive: i2.NgpButton, inputs: ["disabled", "disabled"] }, { directive: i3.NgpFocus, inputs: ["ngpFocusDisabled", "focusDisabled"], outputs: ["ngpFocus", "ngpFocus"] }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: SidebarItem, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: `
    a[flowbiteSidebarItem],
    button[flowbiteSidebarItem]
  `,
                    exportAs: 'flowbiteSidebarItem',
                    hostDirectives: [
                        {
                            directive: BaseButton,
                            inputs: [],
                            outputs: [],
                        },
                        {
                            directive: NgpButton,
                            inputs: ['disabled:disabled'],
                            outputs: [],
                        },
                        {
                            directive: NgpFocus,
                            inputs: ['ngpFocusDisabled:focusDisabled'],
                            outputs: ['ngpFocus'],
                        },
                    ],
                    providers: [provideFlowbiteSidebarItemState()],
                    host: {
                        '[class]': `theme().host.root`,
                        '(click)': 'onClick()',
                    },
                }]
        }] });

/* Sidebar */

/**
 * Generated bundle index. Do not edit.
 */

export { FlowbiteSidebarConfigToken, FlowbiteSidebarContentConfigToken, FlowbiteSidebarContentStateToken, FlowbiteSidebarItemConfigToken, FlowbiteSidebarItemStateToken, FlowbiteSidebarStateToken, FlowbiteSidebarToggleConfigToken, FlowbiteSidebarToggleStateToken, Sidebar, SidebarContent, SidebarItem, SidebarToggle, defaultFlowbiteSidebarConfig, defaultFlowbiteSidebarContentConfig, defaultFlowbiteSidebarItemConfig, defaultFlowbiteSidebarToggleConfig, flowbiteSidebarContentState, flowbiteSidebarContentTheme, flowbiteSidebarItemState, flowbiteSidebarItemTheme, flowbiteSidebarState, flowbiteSidebarTheme, flowbiteSidebarToggleState, flowbiteSidebarToggleTheme, injectFlowbiteSidebarConfig, injectFlowbiteSidebarContentConfig, injectFlowbiteSidebarContentState, injectFlowbiteSidebarItemConfig, injectFlowbiteSidebarItemState, injectFlowbiteSidebarState, injectFlowbiteSidebarToggleConfig, injectFlowbiteSidebarToggleState, provideFlowbiteSidebarConfig, provideFlowbiteSidebarContentConfig, provideFlowbiteSidebarContentState, provideFlowbiteSidebarItemConfig, provideFlowbiteSidebarItemState, provideFlowbiteSidebarState, provideFlowbiteSidebarToggleConfig, provideFlowbiteSidebarToggleState };
