// flowbite-angular/fesm2022/flowbite-angular-theme-toggle.mjs
import { createTheme, mergeDeep, colorToTheme } from 'flowbite-angular';
import * as i0 from '@angular/core';
import { InjectionToken, inject, input, afterNextRender, untracked, Directive, computed, ChangeDetectionStrategy, ViewEncapsulation, Component } from '@angular/core';
import { createStateToken, createStateProvider, createStateInjector, createState } from 'ng-primitives/state';
import * as i1 from 'flowbite-angular/button';
import { injectFlowbiteBaseButtonState, BaseButton } from 'flowbite-angular/button';
import { Icon } from 'flowbite-angular/icon';
import { moon, sun } from 'flowbite-angular/icon/outline/weather';
import { provideIcons } from '@ng-icons/core';
import * as i2 from 'ng-primitives/button';
import { NgpButton } from 'ng-primitives/button';
import * as i3 from 'ng-primitives/interactions';
import { NgpFocus } from 'ng-primitives/interactions';
import { twMerge } from 'tailwind-merge';

const flowbiteThemeToggleTheme = createTheme({
    host: {
        base: 'inline-flex cursor-pointer items-center justify-center rounded-lg p-2 text-sm',
        transition: '',
        focus: 'data-focus:ring-0 data-focus:outline-none data-focus-visible:ring-2 data-focus-visible:outline-none',
        disabled: 'data-disabled:cursor-not-allowed data-disabled:opacity-50',
        size: {},
        pill: {},
        color: {
            default: {
                light: 'text-gray-800 data-hover:text-gray-900',
                dark: 'dark:text-white dark:data-hover:text-gray-100',
            },
            info: {
                light: 'text-gray-800 data-hover:text-blue-900',
                dark: 'dark:text-white dark:data-hover:text-blue-100',
            },
            failure: {
                light: 'text-gray-800 data-hover:text-red-900',
                dark: 'dark:text-white dark:data-hover:text-red-100',
            },
            success: {
                light: 'text-gray-800 data-hover:text-green-900',
                dark: 'dark:text-white dark:data-hover:text-green-100',
            },
            warning: {
                light: 'text-gray-800 data-hover:text-yellow-900',
                dark: 'dark:text-white dark:data-hover:text-yellow-100',
            },
            primary: {
                light: 'data-hover:text-primary-900 text-gray-800',
                dark: 'dark:data-hover:text-primary-100 dark:text-white',
            },
        },
        colorOutline: {},
    },
});

const defaultFlowbiteThemeToggleConfig = {
    baseTheme: flowbiteThemeToggleTheme,
    customTheme: {},
};
const FlowbiteThemeToggleConfigToken = new InjectionToken('FlowbiteThemeToggleConfigToken');
/**
 * Provide the default ThemeToggle configuration
 * @param config The ThemeToggle configuration
 * @returns The provider
 */
const provideFlowbiteThemeToggleConfig = (config) => [
    {
        provide: FlowbiteThemeToggleConfigToken,
        useValue: { ...defaultFlowbiteThemeToggleConfig, ...config },
    },
];
/**
 * Inject the ThemeToggle configuration
 * @see {@link defaultFlowbiteThemeToggleConfig}
 * @returns The configuration
 */
const injectFlowbiteThemeToggleConfig = () => inject(FlowbiteThemeToggleConfigToken, { optional: true }) ?? defaultFlowbiteThemeToggleConfig;

const defaultFlowbiteThemeConfig = {
    theme: 'light',
    localStorageKey: 'fba-color-theme',
    type: { type: 'class' },
};
const FlowbiteThemeConfigToken = new InjectionToken('FlowbiteThemeConfigToken');
/**
 * Provide the default Theme configuration
 * @param config The Theme configuration
 * @returns The provider
 */
const provideFlowbiteThemeConfig = (config) => [
    {
        provide: FlowbiteThemeConfigToken,
        useValue: { ...defaultFlowbiteThemeConfig, ...config },
    },
];
/**
 * Inject the Theme configuration
 * @see {@link defaultFlowbiteThemeConfig}
 * @returns The configuration
 */
const injectFlowbiteThemeConfig = () => inject(FlowbiteThemeConfigToken, { optional: true }) ?? defaultFlowbiteThemeConfig;

const FlowbiteThemeStateToken = createStateToken('Flowbite Theme');
const provideFlowbiteThemeState = createStateProvider(FlowbiteThemeStateToken);
const injectFlowbiteThemeState = createStateInjector(FlowbiteThemeStateToken);
const flowbiteThemeState = createState(FlowbiteThemeStateToken);

class Theme {
    constructor() {
        this.config = injectFlowbiteThemeConfig();
        this.localStorageKey = input(this.config.localStorageKey);
        this.currentTheme = input(this.config.theme);
        this.type = input(this.config.type);
        /**
         * @internal
         */
        this.state = flowbiteThemeState(this);
        afterNextRender(() => {
            this.toggleTheme(this.getLocalStorageTheme());
        });
    }
    getLocalStorageTheme() {
        const localStorageValue = localStorage.getItem(this.state.localStorageKey());
        if (localStorageValue === undefined ||
            (localStorageValue !== 'light' && localStorageValue !== 'dark')) {
            return undefined;
        }
        return localStorageValue;
    }
    toggleTheme(theme) {
        if (!theme) {
            theme = untracked(() => (this.state.currentTheme() === 'light' ? 'dark' : 'light'));
        }
        this.setTheme(theme);
    }
    setTheme(theme) {
        try {
            localStorage.setItem(this.state.localStorageKey(), theme);
        }
        catch (error) {
            console.warn('Failed to save theme to localStorage:', error);
        }
        this.state.currentTheme.set(theme);
        const type = untracked(() => this.state.type());
        if (type.type === 'class') {
            theme === 'light'
                ? document.documentElement.classList.remove('dark')
                : document.documentElement.classList.add('dark');
        }
        else if (type.type === 'attr') {
            document.documentElement.setAttribute(type.name, theme);
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: Theme, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.0.6", type: Theme, isStandalone: true, selector: "\n    [flowbiteTheme]\n  ", inputs: { localStorageKey: { classPropertyName: "localStorageKey", publicName: "localStorageKey", isSignal: true, isRequired: false, transformFunction: null }, currentTheme: { classPropertyName: "currentTheme", publicName: "currentTheme", isSignal: true, isRequired: false, transformFunction: null }, type: { classPropertyName: "type", publicName: "type", isSignal: true, isRequired: false, transformFunction: null } }, providers: [provideFlowbiteThemeState()], exportAs: ["flowbiteTheme"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: Theme, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: `
    [flowbiteTheme]
  `,
                    exportAs: 'flowbiteTheme',
                    hostDirectives: [],
                    providers: [provideFlowbiteThemeState()],
                    host: {},
                }]
        }], ctorParameters: () => [] });

const FlowbiteThemeToggleStateToken = createStateToken('Flowbite ThemeToggle');
const provideFlowbiteThemeToggleState = createStateProvider(FlowbiteThemeToggleStateToken);
const injectFlowbiteThemeToggleState = createStateInjector(FlowbiteThemeToggleStateToken);
const flowbiteThemeToggleState = createState(FlowbiteThemeToggleStateToken);

class ThemeToggle {
    constructor() {
        this.baseButtonState = injectFlowbiteBaseButtonState();
        this.config = injectFlowbiteThemeToggleConfig();
        this.themeDirective = inject(Theme);
        /**
         * @see {@link injectFlowbiteThemeToggleConfig}
         */
        this.customTheme = input(this.config.customTheme);
        this.theme = computed(() => {
            const mergedTheme = mergeDeep(this.config.baseTheme, this.state.customTheme());
            return {
                host: {
                    root: twMerge(mergedTheme.host.base, mergedTheme.host.transition, mergedTheme.host.focus, mergedTheme.host.disabled, colorToTheme(mergedTheme.host.color, this.baseButtonState().color())),
                },
            };
        });
        /**
         * @internal
         */
        this.state = flowbiteThemeToggleState(this);
    }
    /**
     * @internal
     */
    onClick() {
        this.themeDirective.toggleTheme();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: ThemeToggle, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.1.0", version: "20.0.6", type: ThemeToggle, isStandalone: true, selector: "\n    button[flowbiteThemeToggle]\n  ", inputs: { customTheme: { classPropertyName: "customTheme", publicName: "customTheme", isSignal: true, isRequired: false, transformFunction: null } }, host: { listeners: { "click": "onClick()" }, properties: { "class": "theme().host.root" } }, providers: [provideFlowbiteThemeToggleState(), provideIcons({ sun, moon })], exportAs: ["flowbiteThemeToggle"], hostDirectives: [{ directive: i1.BaseButton, inputs: ["color", "color"] }, { directive: i2.NgpButton, inputs: ["disabled", "disabled"] }, { directive: i3.NgpFocus, inputs: ["ngpFocusDisabled", "focusDisabled"], outputs: ["ngpFocus", "ngpFocus"] }], ngImport: i0, template: `
    <flowbite-icon
      class="inline size-5 stroke-2 dark:hidden"
      name="sun" />
    <flowbite-icon
      class="hidden size-5 stroke-2 dark:inline"
      name="moon" />
  `, isInline: true, dependencies: [{ kind: "component", type: Icon, selector: "flowbite-icon", inputs: ["name", "svg", "color", "customTheme"], exportAs: ["flowbiteIcon"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: ThemeToggle, decorators: [{
            type: Component,
            args: [{
                    standalone: true,
                    selector: `
    button[flowbiteThemeToggle]
  `,
                    exportAs: 'flowbiteThemeToggle',
                    hostDirectives: [
                        {
                            directive: BaseButton,
                            inputs: ['color:color'],
                            outputs: [],
                        },
                        {
                            directive: NgpButton,
                            inputs: ['disabled:disabled'],
                            outputs: [],
                        },
                        {
                            directive: NgpFocus,
                            inputs: ['ngpFocusDisabled:focusDisabled'],
                            outputs: ['ngpFocus'],
                        },
                    ],
                    imports: [Icon],
                    providers: [provideFlowbiteThemeToggleState(), provideIcons({ sun, moon })],
                    host: {
                        '[class]': `theme().host.root`,
                        '(click)': 'onClick()',
                    },
                    template: `
    <flowbite-icon
      class="inline size-5 stroke-2 dark:hidden"
      name="sun" />
    <flowbite-icon
      class="hidden size-5 stroke-2 dark:inline"
      name="moon" />
  `,
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }] });

/* ThemeToggle */

/**
 * Generated bundle index. Do not edit.
 */

export { FlowbiteThemeConfigToken, FlowbiteThemeStateToken, FlowbiteThemeToggleConfigToken, FlowbiteThemeToggleStateToken, Theme, ThemeToggle, defaultFlowbiteThemeConfig, defaultFlowbiteThemeToggleConfig, flowbiteThemeState, flowbiteThemeToggleState, flowbiteThemeToggleTheme, injectFlowbiteThemeConfig, injectFlowbiteThemeState, injectFlowbiteThemeToggleConfig, injectFlowbiteThemeToggleState, provideFlowbiteThemeConfig, provideFlowbiteThemeState, provideFlowbiteThemeToggleConfig, provideFlowbiteThemeToggleState };
