// flowbite-angular/fesm2022/flowbite-angular-alert.mjs
import { createTheme, mergeDeep, colorToTheme } from 'flowbite-angular';
import * as i0 from '@angular/core';
import { InjectionToken, inject, input, booleanAttribute, computed, Directive } from '@angular/core';
import { createStateToken, createStateProvider, createStateInjector, createState } from 'ng-primitives/state';
import { twMerge } from 'tailwind-merge';
import * as i1 from 'ng-primitives/button';
import { NgpButton } from 'ng-primitives/button';
import * as i2 from 'ng-primitives/interactions';
import { NgpFocus } from 'ng-primitives/interactions';

const flowbiteAlertTheme = createTheme({
    host: {
        base: 'mb-4 rounded-lg p-4',
        transition: '',
        color: {
            default: {
                light: 'border-gray-200 bg-gray-50',
                dark: 'dark:border-gray-700 dark:bg-gray-950',
            },
            info: {
                light: 'border-blue-200 bg-blue-50',
                dark: 'dark:border-blue-700 dark:bg-blue-950',
            },
            failure: {
                light: 'border-red-200 bg-red-50',
                dark: 'dark:border-red-700 dark:bg-red-950',
            },
            success: {
                light: 'border-green-200 bg-green-50',
                dark: 'dark:border-green-700 dark:bg-green-950',
            },
            warning: {
                light: 'border-yellow-200 bg-yellow-50',
                dark: 'dark:border-yellow-700 dark:bg-yellow-950',
            },
            primary: {
                light: 'bg-primary-50 border-primary-200',
                dark: 'dark:bg-primary-950 dark:border-primary-700',
            },
        },
        border: {
            on: 'border',
            off: 'border-0',
        },
        accent: {
            on: 'border-t-4',
            off: '',
        },
    },
});

const defaultFlowbiteAlertConfig = {
    baseTheme: flowbiteAlertTheme,
    color: 'default',
    border: false,
    accent: false,
    customTheme: {},
};
const FlowbiteAlertConfigToken = new InjectionToken('FlowbiteAlertConfigToken');
/**
 * Provide the default Alert configuration
 * @param config The Alert configuration
 * @returns The provider
 */
const provideFlowbiteAlertConfig = (config) => [
    {
        provide: FlowbiteAlertConfigToken,
        useValue: { ...defaultFlowbiteAlertConfig, ...config },
    },
];
/**
 * Inject the Alert configuration
 * @see {@link defaultFlowbiteAlertConfig}
 * @returns The configuration
 */
const injectFlowbiteAlertConfig = () => inject(FlowbiteAlertConfigToken, { optional: true }) ?? defaultFlowbiteAlertConfig;

const FlowbiteAlertStateToken = createStateToken('Flowbite Alert');
const provideFlowbiteAlertState = createStateProvider(FlowbiteAlertStateToken);
const injectFlowbiteAlertState = createStateInjector(FlowbiteAlertStateToken);
const flowbiteAlertState = createState(FlowbiteAlertStateToken);

class Alert {
    constructor() {
        this.config = injectFlowbiteAlertConfig();
        /**
         * @see {@link injectFlowbiteAlertConfig}
         */
        this.color = input(this.config.color);
        /**
         * @see {@link injectFlowbiteAlertConfig}
         */
        this.border = input(this.config.border, { transform: booleanAttribute });
        /**
         * @see {@link injectFlowbiteAlertConfig}
         */
        this.accent = input(this.config.accent, { transform: booleanAttribute });
        /**
         * @see {@link injectFlowbiteAlertConfig}
         */
        this.customTheme = input(this.config.customTheme);
        this.theme = computed(() => {
            const mergedTheme = mergeDeep(this.config.baseTheme, this.state.customTheme());
            return {
                host: {
                    root: twMerge(mergedTheme.host.base, mergedTheme.host.transition, colorToTheme(mergedTheme.host.color, this.state.color()), mergedTheme.host.border[this.state.border() ? 'on' : 'off'], mergedTheme.host.accent[this.state.accent() ? 'on' : 'off']),
                },
            };
        });
        /**
         * @internal
         */
        this.state = flowbiteAlertState(this);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: Alert, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.0.6", type: Alert, isStandalone: true, selector: "\n    div[flowbiteAlert]\n  ", inputs: { color: { classPropertyName: "color", publicName: "color", isSignal: true, isRequired: false, transformFunction: null }, border: { classPropertyName: "border", publicName: "border", isSignal: true, isRequired: false, transformFunction: null }, accent: { classPropertyName: "accent", publicName: "accent", isSignal: true, isRequired: false, transformFunction: null }, customTheme: { classPropertyName: "customTheme", publicName: "customTheme", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "class": "theme().host.root" } }, providers: [provideFlowbiteAlertState()], exportAs: ["flowbiteAlert"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: Alert, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: `
    div[flowbiteAlert]
  `,
                    exportAs: 'flowbiteAlert',
                    hostDirectives: [],
                    providers: [provideFlowbiteAlertState()],
                    host: {
                        '[class]': `theme().host.root`,
                    },
                }]
        }] });

const flowbiteAlertButtonTheme = createTheme({
    host: {
        base: 'flex rounded-lg p-1 first:mr-2 data-hover:cursor-pointer',
        transition: '',
        color: {
            default: {
                light: 'data-hover:bg-gray-300',
                dark: 'dark:data-hover:bg-gray-600',
            },
            info: {
                light: 'data-hover:bg-blue-300',
                dark: 'dark:data-hover:bg-blue-700',
            },
            failure: {
                light: 'data-hover:bg-red-300',
                dark: 'dark:data-hover:bg-red-700',
            },
            success: {
                light: 'data-hover:bg-green-300',
                dark: 'dark:data-hover:bg-green-600',
            },
            warning: {
                light: 'data-hover:bg-yellow-300',
                dark: 'dark:data-hover:bg-yellow-600',
            },
            primary: {
                light: 'data-hover:bg-primary-300',
                dark: 'dark:data-hover:bg-primary-700',
            },
        },
    },
});

const defaultFlowbiteAlertButtonConfig = {
    baseTheme: flowbiteAlertButtonTheme,
    customTheme: {},
};
const FlowbiteAlertButtonConfigToken = new InjectionToken('FlowbiteAlertButtonConfigToken');
/**
 * Provide the default AlertButton configuration
 * @param config The AlertButton configuration
 * @returns The provider
 */
const provideFlowbiteAlertButtonConfig = (config) => [
    {
        provide: FlowbiteAlertButtonConfigToken,
        useValue: { ...defaultFlowbiteAlertButtonConfig, ...config },
    },
];
/**
 * Inject the AlertButton configuration
 * @see {@link defaultFlowbiteAlertButtonConfig}
 * @returns The configuration
 */
const injectFlowbiteAlertButtonConfig = () => inject(FlowbiteAlertButtonConfigToken, { optional: true }) ?? defaultFlowbiteAlertButtonConfig;

const FlowbiteAlertButtonStateToken = createStateToken('Flowbite AlertButton');
const provideFlowbiteAlertButtonState = createStateProvider(FlowbiteAlertButtonStateToken);
const injectFlowbiteAlertButtonState = createStateInjector(FlowbiteAlertButtonStateToken);
const flowbiteAlertButtonState = createState(FlowbiteAlertButtonStateToken);

class AlertButton {
    constructor() {
        this.config = injectFlowbiteAlertButtonConfig();
        this.alertState = injectFlowbiteAlertState();
        /**
         * @see {@link injectFlowbiteAlertButtonConfig}
         */
        this.customTheme = input(this.config.customTheme);
        this.theme = computed(() => {
            const mergedTheme = mergeDeep(this.config.baseTheme, this.state.customTheme());
            return {
                host: {
                    root: twMerge(mergedTheme.host.base, mergedTheme.host.transition, colorToTheme(mergedTheme.host.color, this.alertState().color())),
                },
            };
        });
        /**
         * @internal
         */
        this.state = flowbiteAlertButtonState(this);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: AlertButton, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.0.6", type: AlertButton, isStandalone: true, selector: "\n    button[flowbiteAlertButton],\n    a[flowbiteAlertButton]\n  ", inputs: { customTheme: { classPropertyName: "customTheme", publicName: "customTheme", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "class": "theme().host.root" } }, providers: [provideFlowbiteAlertButtonState()], exportAs: ["flowbiteAlertButton"], hostDirectives: [{ directive: i1.NgpButton, inputs: ["disabled", "disabled"] }, { directive: i2.NgpFocus }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: AlertButton, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: `
    button[flowbiteAlertButton],
    a[flowbiteAlertButton]
  `,
                    exportAs: 'flowbiteAlertButton',
                    hostDirectives: [
                        {
                            directive: NgpButton,
                            inputs: ['disabled:disabled'],
                            outputs: [],
                        },
                        {
                            directive: NgpFocus,
                            inputs: [],
                            outputs: [],
                        },
                    ],
                    providers: [provideFlowbiteAlertButtonState()],
                    host: { '[class]': `theme().host.root` },
                }]
        }] });

const flowbiteAlertContentTheme = createTheme({
    host: {
        base: 'flex items-center justify-between text-sm font-normal',
        transition: '',
        color: {
            default: {
                light: 'text-gray-800',
                dark: 'dark:text-gray-300',
            },
            info: {
                light: 'text-blue-800',
                dark: 'dark:text-blue-300',
            },
            failure: {
                light: 'text-red-800',
                dark: 'dark:text-red-300',
            },
            success: {
                light: 'text-green-800',
                dark: 'dark:text-green-300',
            },
            warning: {
                light: 'text-yellow-800',
                dark: 'dark:text-yellow-300',
            },
            primary: {
                light: 'text-primary-800',
                dark: 'dark:text-primary-300',
            },
        },
    },
});

const defaultFlowbiteAlertContentConfig = {
    baseTheme: flowbiteAlertContentTheme,
    customTheme: {},
};
const FlowbiteAlertContentConfigToken = new InjectionToken('FlowbiteAlertContentConfigToken');
/**
 * Provide the default AlertContent configuration
 * @param config The AlertContent configuration
 * @returns The provider
 */
const provideFlowbiteAlertContentConfig = (config) => [
    {
        provide: FlowbiteAlertContentConfigToken,
        useValue: { ...defaultFlowbiteAlertContentConfig, ...config },
    },
];
/**
 * Inject the AlertContent configuration
 * @see {@link defaultFlowbiteAlertContentConfig}
 * @returns The configuration
 */
const injectFlowbiteAlertContentConfig = () => inject(FlowbiteAlertContentConfigToken, { optional: true }) ?? defaultFlowbiteAlertContentConfig;

const FlowbiteAlertContentStateToken = createStateToken('Flowbite AlertContent');
const provideFlowbiteAlertContentState = createStateProvider(FlowbiteAlertContentStateToken);
const injectFlowbiteAlertContentState = createStateInjector(FlowbiteAlertContentStateToken);
const flowbiteAlertContentState = createState(FlowbiteAlertContentStateToken);

class AlertContent {
    constructor() {
        this.config = injectFlowbiteAlertContentConfig();
        this.alertState = injectFlowbiteAlertState();
        /**
         * @see {@link injectFlowbiteAlertContentConfig}
         */
        this.customTheme = input(this.config.customTheme);
        this.theme = computed(() => {
            const mergedTheme = mergeDeep(this.config.baseTheme, this.state.customTheme());
            return {
                host: {
                    root: twMerge(mergedTheme.host.base, mergedTheme.host.transition, colorToTheme(mergedTheme.host.color, this.alertState().color())),
                },
            };
        });
        /**
         * @internal
         */
        this.state = flowbiteAlertContentState(this);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: AlertContent, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.0.6", type: AlertContent, isStandalone: true, selector: "\n    div[flowbiteAlertContent]\n  ", inputs: { customTheme: { classPropertyName: "customTheme", publicName: "customTheme", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "class": "theme().host.root" } }, providers: [provideFlowbiteAlertContentState()], exportAs: ["flowbiteAlertContent"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.0.6", ngImport: i0, type: AlertContent, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: `
    div[flowbiteAlertContent]
  `,
                    exportAs: 'flowbiteAlertContent',
                    hostDirectives: [],
                    providers: [provideFlowbiteAlertContentState()],
                    host: { '[class]': `theme().host.root` },
                }]
        }] });

/* Alert */

/**
 * Generated bundle index. Do not edit.
 */

export { Alert, AlertButton, AlertContent, FlowbiteAlertButtonConfigToken, FlowbiteAlertButtonStateToken, FlowbiteAlertConfigToken, FlowbiteAlertContentConfigToken, FlowbiteAlertContentStateToken, FlowbiteAlertStateToken, defaultFlowbiteAlertButtonConfig, defaultFlowbiteAlertConfig, defaultFlowbiteAlertContentConfig, flowbiteAlertButtonState, flowbiteAlertButtonTheme, flowbiteAlertContentState, flowbiteAlertContentTheme, flowbiteAlertState, flowbiteAlertTheme, injectFlowbiteAlertButtonConfig, injectFlowbiteAlertButtonState, injectFlowbiteAlertConfig, injectFlowbiteAlertContentConfig, injectFlowbiteAlertContentState, injectFlowbiteAlertState, provideFlowbiteAlertButtonConfig, provideFlowbiteAlertButtonState, provideFlowbiteAlertConfig, provideFlowbiteAlertContentConfig, provideFlowbiteAlertContentState, provideFlowbiteAlertState };
