// ng-primitives/schematics/ng-generate/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = generatePrimitive;
const core_1 = require("@angular-devkit/core");
const schematics_1 = require("@angular-devkit/schematics");
const tsquery_1 = require("@phenomnomnominal/tsquery");
function generatePrimitive(options) {
    return () => {
        const templateSource = (0, schematics_1.apply)((0, schematics_1.url)(`./templates/${options.primitive}`), [
            (0, schematics_1.template)({
                ...options,
                ...core_1.strings,
                prefix: options.prefix ?? 'app',
                fileSuffix: options.fileSuffix ?? 'component',
            }),
            (0, schematics_1.forEach)((fileEntry) => {
                if (fileEntry.path.endsWith('.template')) {
                    let newPath = fileEntry.path.replace('.template', '');
                    // if the file has two consecutive periods, replace them with a single period - this can happen if the fileSuffix is empty
                    if (newPath.includes('..')) {
                        newPath = newPath.replace('..', '.');
                    }
                    return {
                        path: newPath,
                        content: processStyles(fileEntry.content, options),
                    };
                }
                return fileEntry;
            }),
            (0, schematics_1.move)(options.path),
        ]);
        return (0, schematics_1.mergeWith)(templateSource);
    };
}
function processStyles(content, options) {
    if (options.exampleStyles) {
        return content;
    }
    const contentStr = content.toString();
    const styles = (0, tsquery_1.query)(contentStr, 'ClassDeclaration > Decorator > CallExpression:has(Identifier[name="Component"]) ObjectLiteralExpression PropertyAssignment:has(Identifier[name="styles"]) NoSubstitutionTemplateLiteral');
    if (styles.length === 0) {
        return content;
    }
    const stylesNode = styles[0];
    const stylesText = stylesNode.getText();
    const stylesValue = stylesText.slice(1, stylesText.length - 1);
    // we want to preserve all the selectors, we just want to remove all the rules inside the selectors
    const stylesWithoutRules = stylesValue.replace(/(?<=\{)[^}]+(?=\})/g, '');
    // replace the styles value with the new value
    const contentStrWithoutRules = contentStr.replace(stylesValue, stylesWithoutRules);
    // convert back to a buffer
    return Buffer.from(contentStrWithoutRules);
}
