// ng-primitives/fesm2022/ng-primitives-separator.mjs
import * as i0 from '@angular/core';
import { InjectionToken, inject, input, Directive } from '@angular/core';

const defaultSeparatorConfig = {
    orientation: 'horizontal',
};
const NgpSeparatorConfigToken = new InjectionToken('NgpSeparatorConfigToken');
/**
 * Provide the default Separator configuration
 * @param config The Separator configuration
 * @returns The provider
 */
function provideSeparatorConfig(config) {
    return [
        {
            provide: NgpSeparatorConfigToken,
            useValue: { ...defaultSeparatorConfig, ...config },
        },
    ];
}
/**
 * Inject the Separator configuration
 * @returns The global Separator configuration
 */
function injectSeparatorConfig() {
    return inject(NgpSeparatorConfigToken, { optional: true }) ?? defaultSeparatorConfig;
}

class NgpSeparator {
    constructor() {
        this.config = injectSeparatorConfig();
        /**
         * The orientation of the separator.
         */
        this.orientation = input(this.config.orientation, {
            alias: 'ngpSeparatorOrientation',
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpSeparator, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "19.2.11", type: NgpSeparator, isStandalone: true, selector: "[ngpSeparator]", inputs: { orientation: { classPropertyName: "orientation", publicName: "ngpSeparatorOrientation", isSignal: true, isRequired: false, transformFunction: null } }, host: { attributes: { "role": "separator" }, properties: { "attr.aria-orientation": "orientation()", "attr.data-orientation": "orientation()" } }, exportAs: ["ngpSeparator"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpSeparator, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpSeparator]',
                    exportAs: 'ngpSeparator',
                    host: {
                        role: 'separator',
                        '[attr.aria-orientation]': 'orientation()',
                        '[attr.data-orientation]': 'orientation()',
                    },
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { NgpSeparator, provideSeparatorConfig };
