// ng-primitives/fesm2022/ng-primitives-common.mjs
import * as i0 from '@angular/core';
import { InjectionToken, inject, input, Directive } from '@angular/core';
import { uniqueId } from 'ng-primitives/utils';

/**
 * @deprecated
 */
const NgpHeaderToken = new InjectionToken('NgpHeaderToken');
/**
 * Inject the Header directive instance
 */
function injectHeader() {
    return inject(NgpHeaderToken);
}

/**
 * @deprecated
 */
class NgpHeader {
    constructor() {
        /**
         * The id of the header.
         */
        this.id = input(uniqueId('ngp-header'));
        if (ngDevMode) {
            console.warn(`NgpHeader is deprecated and will be removed in a future version. Please use NgpListboxHeader instead.`);
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpHeader, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "19.2.11", type: NgpHeader, isStandalone: true, selector: "[ngpHeader]", inputs: { id: { classPropertyName: "id", publicName: "id", isSignal: true, isRequired: false, transformFunction: null } }, host: { attributes: { "role": "presentation" }, properties: { "attr.id": "id()" } }, providers: [{ provide: NgpHeaderToken, useExisting: NgpHeader }], exportAs: ["ngpHeader"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpHeader, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpHeader]',
                    exportAs: 'ngpHeader',
                    providers: [{ provide: NgpHeaderToken, useExisting: NgpHeader }],
                    host: {
                        role: 'presentation',
                        '[attr.id]': 'id()',
                    },
                }]
        }], ctorParameters: () => [] });

/**
 * Generated bundle index. Do not edit.
 */

export { NgpHeader, NgpHeaderToken, injectHeader };
