// ng-primitives/fesm2022/ng-primitives-toolbar.mjs
import * as i0 from '@angular/core';
import { input, Directive } from '@angular/core';
import { explicitEffect } from 'ng-primitives/internal';
import * as i1 from 'ng-primitives/roving-focus';
import { injectRovingFocusGroupState, NgpRovingFocusGroup } from 'ng-primitives/roving-focus';
import { createStateToken, createStateProvider, createStateInjector, createState } from 'ng-primitives/state';

/**
 * The state token  for the Toolbar primitive.
 */
const NgpToolbarStateToken = createStateToken('Toolbar');
/**
 * Provides the Toolbar state.
 */
const provideToolbarState = createStateProvider(NgpToolbarStateToken);
/**
 * Injects the Toolbar state.
 */
const injectToolbarState = createStateInjector(NgpToolbarStateToken);
/**
 * The Toolbar state registration function.
 */
const toolbarState = createState(NgpToolbarStateToken);

class NgpToolbar {
    constructor() {
        /**
         * Access the roving focus group state.
         */
        this.rovingFocusGroup = injectRovingFocusGroupState();
        /**
         * The orientation of the toolbar.
         */
        this.orientation = input('horizontal', {
            alias: 'ngpToolbarOrientation',
        });
        /**
         * The toolbar state.
         */
        this.state = toolbarState(this);
        explicitEffect([this.state.orientation], ([orientation]) => this.rovingFocusGroup().orientation.set(orientation));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpToolbar, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "19.2.11", type: NgpToolbar, isStandalone: true, selector: "[ngpToolbar]", inputs: { orientation: { classPropertyName: "orientation", publicName: "ngpToolbarOrientation", isSignal: true, isRequired: false, transformFunction: null } }, host: { attributes: { "role": "toolbar" }, properties: { "attr.aria-orientation": "state.orientation()", "attr.data-orientation": "state.orientation()" } }, providers: [provideToolbarState()], exportAs: ["ngpToolbar"], hostDirectives: [{ directive: i1.NgpRovingFocusGroup }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpToolbar, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpToolbar]',
                    exportAs: 'ngpToolbar',
                    providers: [provideToolbarState()],
                    hostDirectives: [NgpRovingFocusGroup],
                    host: {
                        role: 'toolbar',
                        '[attr.aria-orientation]': 'state.orientation()',
                        '[attr.data-orientation]': 'state.orientation()',
                    },
                }]
        }], ctorParameters: () => [] });

/**
 * Generated bundle index. Do not edit.
 */

export { NgpToolbar, injectToolbarState, provideToolbarState };
