// ng-primitives/fesm2022/ng-primitives-checkbox.mjs
import * as i0 from '@angular/core';
import { input, booleanAttribute, output, HostListener, Directive } from '@angular/core';
import { setupFormControl } from 'ng-primitives/form-field';
import { ngpInteractions } from 'ng-primitives/interactions';
import { uniqueId } from 'ng-primitives/utils';
import { createStateToken, createStateProvider, createStateInjector, createState } from 'ng-primitives/state';

/**
 * The state token  for the Checkbox primitive.
 */
const NgpCheckboxStateToken = createStateToken('Checkbox');
/**
 * Provides the Checkbox state.
 */
const provideCheckboxState = createStateProvider(NgpCheckboxStateToken);
/**
 * Injects the Checkbox state.
 */
const injectCheckboxState = createStateInjector(NgpCheckboxStateToken);
/**
 * The Checkbox state registration function.
 */
const checkboxState = createState(NgpCheckboxStateToken);

/**
 * Apply the `ngpCheckbox` directive to an element to that represents the checkbox, such as a `button`.
 */
class NgpCheckbox {
    constructor() {
        /**
         * The id of the checkbox.
         * @internal
         */
        this.id = input(uniqueId('ngp-checkbox'));
        /**
         * Defines whether the checkbox is checked.
         */
        this.checked = input(false, {
            alias: 'ngpCheckboxChecked',
            transform: booleanAttribute,
        });
        /**
         * The event that is emitted when the checkbox value changes.
         */
        this.checkedChange = output({
            alias: 'ngpCheckboxCheckedChange',
        });
        /**
         * Defines whether the checkbox is indeterminate.
         */
        this.indeterminate = input(false, {
            alias: 'ngpCheckboxIndeterminate',
            transform: booleanAttribute,
        });
        /**
         * The event that is emitted when the indeterminate value changes.
         */
        this.indeterminateChange = output({
            alias: 'ngpCheckboxIndeterminateChange',
        });
        /**
         * Whether the checkbox is required.
         */
        this.required = input(false, {
            alias: 'ngpCheckboxRequired',
            transform: booleanAttribute,
        });
        /**
         * Defines whether the checkbox is disabled.
         */
        this.disabled = input(false, {
            alias: 'ngpCheckboxDisabled',
            transform: booleanAttribute,
        });
        /**
         * The state of the checkbox.
         */
        this.state = checkboxState(this);
        setupFormControl({ id: this.state.id, disabled: this.state.disabled });
        ngpInteractions({
            hover: true,
            press: true,
            focusVisible: true,
            disabled: this.state.disabled,
        });
    }
    onEnter(event) {
        // According to WAI ARIA, Checkboxes don't activate on enter keypress
        event.preventDefault();
    }
    toggle(event) {
        if (this.state.disabled()) {
            return;
        }
        // prevent this firing twice in cases where the label is clicked and the checkbox is clicked by the one event
        event?.preventDefault();
        const checked = this.state.indeterminate() ? true : !this.state.checked();
        this.state.checked.set(checked);
        this.checkedChange.emit(checked);
        // if the checkbox was indeterminate, it isn't anymore
        if (this.state.indeterminate()) {
            this.state.indeterminate.set(false);
            this.indeterminateChange.emit(false);
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpCheckbox, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "19.2.11", type: NgpCheckbox, isStandalone: true, selector: "[ngpCheckbox]", inputs: { id: { classPropertyName: "id", publicName: "id", isSignal: true, isRequired: false, transformFunction: null }, checked: { classPropertyName: "checked", publicName: "ngpCheckboxChecked", isSignal: true, isRequired: false, transformFunction: null }, indeterminate: { classPropertyName: "indeterminate", publicName: "ngpCheckboxIndeterminate", isSignal: true, isRequired: false, transformFunction: null }, required: { classPropertyName: "required", publicName: "ngpCheckboxRequired", isSignal: true, isRequired: false, transformFunction: null }, disabled: { classPropertyName: "disabled", publicName: "ngpCheckboxDisabled", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { checkedChange: "ngpCheckboxCheckedChange", indeterminateChange: "ngpCheckboxIndeterminateChange" }, host: { attributes: { "role": "checkbox" }, listeners: { "keydown.enter": "onEnter($event)", "click": "toggle($event)", "keydown.space": "toggle($event)" }, properties: { "attr.aria-checked": "state.indeterminate() ? \"mixed\" : state.checked()", "attr.data-checked": "state.checked() ? \"\" : null", "attr.data-indeterminate": "state.indeterminate() ? \"\" : null", "attr.aria-disabled": "state.disabled()", "tabindex": "state.disabled() ? -1 : 0" } }, providers: [provideCheckboxState()], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpCheckbox, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpCheckbox]',
                    providers: [provideCheckboxState()],
                    host: {
                        role: 'checkbox',
                        '[attr.aria-checked]': 'state.indeterminate() ? "mixed" : state.checked()',
                        '[attr.data-checked]': 'state.checked() ? "" : null',
                        '[attr.data-indeterminate]': 'state.indeterminate() ? "" : null',
                        '[attr.aria-disabled]': 'state.disabled()',
                        '[tabindex]': 'state.disabled() ? -1 : 0',
                    },
                }]
        }], ctorParameters: () => [], propDecorators: { onEnter: [{
                type: HostListener,
                args: ['keydown.enter', ['$event']]
            }], toggle: [{
                type: HostListener,
                args: ['click', ['$event']]
            }, {
                type: HostListener,
                args: ['keydown.space', ['$event']]
            }] } });

/**
 * Generated bundle index. Do not edit.
 */

export { NgpCheckbox, injectCheckboxState, provideCheckboxState };
