// ng-primitives/fesm2022/ng-primitives-progress.mjs
import * as i0 from '@angular/core';
import { computed, Directive, input, numberAttribute, signal } from '@angular/core';
import { createStateToken, createStateProvider, createStateInjector, createState } from 'ng-primitives/state';
import { uniqueId } from 'ng-primitives/utils';

/**
 * The state token  for the Progress primitive.
 */
const NgpProgressStateToken = createStateToken('Progress');
/**
 * Provides the Progress state.
 */
const provideProgressState = createStateProvider(NgpProgressStateToken);
/**
 * Injects the Progress state.
 */
const injectProgressState = createStateInjector(NgpProgressStateToken);
/**
 * The Progress state registration function.
 */
const progressState = createState(NgpProgressStateToken);

/**
 * Apply the `ngpProgressIndicator` directive to an element that represents the current progress.
 * The width of this element can be set to the percentage of the progress value.
 */
class NgpProgressIndicator {
    constructor() {
        /**
         * Access the progress state.
         */
        this.state = injectProgressState();
        /**
         * Get the percentage of the progress value.
         */
        this.percentage = computed(() => this.state().value() === null
            ? null
            : ((this.state().value() - this.state().min()) / (this.state().max() - this.state().min())) *
                100);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpProgressIndicator, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "19.2.11", type: NgpProgressIndicator, isStandalone: true, selector: "[ngpProgressIndicator]", host: { properties: { "style.width.%": "percentage()", "attr.data-progressing": "state().progressing() ? \"\" : null", "attr.data-indeterminate": "state().indeterminate() ? \"\" : null", "attr.data-complete": "state().complete() ? \"\" : null" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpProgressIndicator, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpProgressIndicator]',
                    host: {
                        '[style.width.%]': 'percentage()',
                        '[attr.data-progressing]': 'state().progressing() ? "" : null',
                        '[attr.data-indeterminate]': 'state().indeterminate() ? "" : null',
                        '[attr.data-complete]': 'state().complete() ? "" : null',
                    },
                }]
        }] });

class NgpProgressLabel {
    constructor() {
        /**
         * Access the progress state.
         */
        this.state = injectProgressState();
        /**
         * The unique identifier for the progress label.
         */
        this.id = input(uniqueId('ngp-progress-label'));
        this.state().label.set(this);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpProgressLabel, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "19.2.11", type: NgpProgressLabel, isStandalone: true, selector: "[ngpProgressLabel]", inputs: { id: { classPropertyName: "id", publicName: "id", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "attr.data-progressing": "state().progressing() ? \"\" : null", "attr.data-indeterminate": "state().indeterminate() ? \"\" : null", "attr.data-complete": "state().complete() ? \"\" : null" } }, exportAs: ["ngpProgressLabel"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpProgressLabel, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpProgressLabel]',
                    exportAs: 'ngpProgressLabel',
                    host: {
                        '[attr.data-progressing]': 'state().progressing() ? "" : null',
                        '[attr.data-indeterminate]': 'state().indeterminate() ? "" : null',
                        '[attr.data-complete]': 'state().complete() ? "" : null',
                    },
                }]
        }], ctorParameters: () => [] });

class NgpProgressTrack {
    constructor() {
        /**
         * Access the progress state.
         */
        this.state = injectProgressState();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpProgressTrack, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "19.2.11", type: NgpProgressTrack, isStandalone: true, selector: "[ngpProgressTrack]", host: { properties: { "attr.data-progressing": "state().progressing() ? \"\" : null", "attr.data-indeterminate": "state().indeterminate() ? \"\" : null", "attr.data-complete": "state().complete() ? \"\" : null" } }, exportAs: ["ngpProgressTrack"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpProgressTrack, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpProgressTrack]',
                    exportAs: 'ngpProgressTrack',
                    host: {
                        '[attr.data-progressing]': 'state().progressing() ? "" : null',
                        '[attr.data-indeterminate]': 'state().indeterminate() ? "" : null',
                        '[attr.data-complete]': 'state().complete() ? "" : null',
                    },
                }]
        }] });

class NgpProgressValue {
    constructor() {
        /**
         * Access the progress state.
         */
        this.state = injectProgressState();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpProgressValue, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "19.2.11", type: NgpProgressValue, isStandalone: true, selector: "[ngpProgressValue]", host: { attributes: { "aria-hidden": "true" }, properties: { "attr.data-progressing": "state().progressing() ? \"\" : null", "attr.data-indeterminate": "state().indeterminate() ? \"\" : null", "attr.data-complete": "state().complete() ? \"\" : null" } }, exportAs: ["ngpProgressValue"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpProgressValue, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpProgressValue]',
                    exportAs: 'ngpProgressValue',
                    host: {
                        'aria-hidden': 'true',
                        '[attr.data-progressing]': 'state().progressing() ? "" : null',
                        '[attr.data-indeterminate]': 'state().indeterminate() ? "" : null',
                        '[attr.data-complete]': 'state().complete() ? "" : null',
                    },
                }]
        }] });

/**
 * Apply the `ngpProgress` directive to an element that represents the progress bar.
 */
class NgpProgress {
    constructor() {
        /**
         * Define the progress value.
         */
        this.value = input(0, {
            alias: 'ngpProgressValue',
            transform: v => (v == null ? null : numberAttribute(v)),
        });
        /**
         * Define the progress min value.
         * @default '0'
         */
        this.min = input(0, {
            alias: 'ngpProgressMin',
            transform: numberAttribute,
        });
        /**
         * Define the progress max value.
         * @default 100
         */
        this.max = input(100, {
            alias: 'ngpProgressMax',
            transform: numberAttribute,
        });
        /**
         * Define a function that returns the progress value label.
         * @param value The current value
         * @param max The maximum value
         * @returns The value label
         */
        this.valueLabel = input((value, max) => `${Math.round((value / max) * 100)}%`, {
            alias: 'ngpProgressValueLabel',
        });
        /**
         * Determine if the progress is indeterminate.
         * @internal
         */
        this.indeterminate = computed(() => this.state.value() === null);
        /**
         * Determine if the progress is in a progressing state.
         * @internal
         */
        this.progressing = computed(() => this.state.value() != null &&
            this.state.value() > 0 &&
            this.state.value() < this.state.max());
        /**
         * Determine if the progress is complete.
         * @internal
         */
        this.complete = computed(() => this.state.value() === this.state.max());
        /**
         * Get the progress value text.
         */
        this.valueText = computed(() => {
            const value = this.state.value();
            if (value == null) {
                return '';
            }
            return this.state.valueLabel()(value, this.state.max());
        });
        /**
         * The label associated with the progress bar.
         * @internal
         */
        this.label = signal(null);
        /**
         * The state of the progress bar.
         * @internal
         */
        this.state = progressState(this);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpProgress, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "19.2.11", type: NgpProgress, isStandalone: true, selector: "[ngpProgress]", inputs: { value: { classPropertyName: "value", publicName: "ngpProgressValue", isSignal: true, isRequired: false, transformFunction: null }, min: { classPropertyName: "min", publicName: "ngpProgressMin", isSignal: true, isRequired: false, transformFunction: null }, max: { classPropertyName: "max", publicName: "ngpProgressMax", isSignal: true, isRequired: false, transformFunction: null }, valueLabel: { classPropertyName: "valueLabel", publicName: "ngpProgressValueLabel", isSignal: true, isRequired: false, transformFunction: null } }, host: { attributes: { "role": "progressbar" }, properties: { "attr.aria-valuemax": "state.max()", "attr.aria-valuemin": "0", "attr.aria-valuenow": "state.value()", "attr.aria-valuetext": "valueText()", "attr.aria-labelledby": "label() ? label().id : null", "attr.data-progressing": "progressing() ? \"\" : null", "attr.data-indeterminate": "indeterminate() ? \"\" : null", "attr.data-complete": "complete() ? \"\" : null" } }, providers: [provideProgressState()], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpProgress, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpProgress]',
                    providers: [provideProgressState()],
                    host: {
                        role: 'progressbar',
                        '[attr.aria-valuemax]': 'state.max()',
                        '[attr.aria-valuemin]': '0',
                        '[attr.aria-valuenow]': 'state.value()',
                        '[attr.aria-valuetext]': 'valueText()',
                        '[attr.aria-labelledby]': 'label() ? label().id : null',
                        '[attr.data-progressing]': 'progressing() ? "" : null',
                        '[attr.data-indeterminate]': 'indeterminate() ? "" : null',
                        '[attr.data-complete]': 'complete() ? "" : null',
                    },
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { NgpProgress, NgpProgressIndicator, NgpProgressLabel, NgpProgressTrack, NgpProgressValue, injectProgressState, provideProgressState };
