// ng-primitives/fesm2022/ng-primitives-file-upload.mjs
import { coerceStringArray } from '@angular/cdk/coercion';
import * as i0 from '@angular/core';
import { inject, ElementRef, input, booleanAttribute, output, signal, HostListener, Directive } from '@angular/core';
import { ngpHoverInteraction, ngpInteractions } from 'ng-primitives/interactions';
import { createStateToken, createStateProvider, createStateInjector, createState } from 'ng-primitives/state';
import { DOCUMENT } from '@angular/common';

function fileDropFilter(fileList, acceptedTypes, multiple) {
    const validFiles = Array.from(fileList).filter(file => isFileTypeAccepted(file, acceptedTypes));
    const limitedFiles = multiple ? validFiles : validFiles.slice(0, 1);
    return limitedFiles.length > 0 ? filesToFileList(limitedFiles) : null;
}
function isFileTypeAccepted(file, acceptedTypes) {
    // allow all file types if no types are specified
    if (!acceptedTypes || acceptedTypes.length === 0)
        return true;
    const mimeType = file.type;
    const fileName = file.name.toLowerCase();
    return acceptedTypes.some(type => {
        type = type.toLowerCase();
        if (type.startsWith('.')) {
            return fileName.endsWith(type);
        }
        if (type.endsWith('/*')) {
            const baseType = type.replace('/*', '');
            return mimeType.startsWith(baseType);
        }
        return mimeType === type;
    });
}
function filesToFileList(files) {
    const dataTransfer = new DataTransfer();
    files.forEach(file => dataTransfer.items.add(file));
    return dataTransfer.files;
}

/**
 * The state token  for the FileDropzone primitive.
 */
const NgpFileDropzoneStateToken = createStateToken('FileDropzone');
/**
 * Provides the FileDropzone state.
 */
const provideFileDropzoneState = createStateProvider(NgpFileDropzoneStateToken);
/**
 * Injects the FileDropzone state.
 */
const injectFileDropzoneState = createStateInjector(NgpFileDropzoneStateToken);
/**
 * The FileDropzone state registration function.
 */
const fileDropzoneState = createState(NgpFileDropzoneStateToken);

/**
 * Capture files dropped on the element.
 */
class NgpFileDropzone {
    constructor() {
        /**
         * Access the host element.
         */
        this.elementRef = inject(ElementRef);
        /**
         * The accepted file types. This can be an array of strings or a comma-separated string.
         * Accepted types can either be file extensions (e.g. `.jpg`) or MIME types (e.g. `image/jpeg`).
         */
        this.fileTypes = input(undefined, {
            alias: 'ngpFileDropzoneFileTypes',
            transform: types => coerceStringArray(types, ','),
        });
        /**
         * Whether to allow multiple files to be selected.
         */
        this.multiple = input(false, {
            alias: 'ngpFileDropzoneMultiple',
            transform: booleanAttribute,
        });
        /**
         * Whether to allow the user to select directories.
         */
        this.directory = input(false, {
            alias: 'ngpFileDropzoneDirectory',
            transform: booleanAttribute,
        });
        /**
         * Whether the file upload is disabled.
         */
        this.disabled = input(false, {
            alias: 'ngpFileDropzoneDisabled',
            transform: booleanAttribute,
        });
        /**
         * Emits when the user selects files.
         */
        this.selected = output({
            alias: 'ngpFileDropzoneSelected',
        });
        /**
         * Emits when uploaded files are rejected because they do not match the allowed {@link fileTypes}.
         */
        this.rejected = output({
            alias: 'ngpFileDropzoneRejected',
        });
        /**
         * Emits when the user drags a file over the file upload.
         */
        this.dragOver = output({
            alias: 'ngpFileDropzoneDragOver',
        });
        /**
         * Whether the user is currently dragging a file over the file upload.
         */
        this.isDragOver = signal(false);
        /**
         * The file upload state.
         */
        this.state = fileDropzoneState(this);
        ngpHoverInteraction({ disabled: this.state.disabled });
    }
    onDragEnter(event) {
        if (this.state.disabled()) {
            return;
        }
        event.preventDefault();
        event.stopPropagation();
        this.isDragOver.set(true);
        this.dragOver.emit(true);
    }
    onDragOver(event) {
        if (this.state.disabled()) {
            return;
        }
        event.stopPropagation();
        event.preventDefault();
        this.isDragOver.set(true);
    }
    onDragLeave(event) {
        if (this.state.disabled() || !this.isDragOver()) {
            return;
        }
        // if the element we are dragging over is a child of the file upload, ignore the event
        if (this.elementRef.nativeElement.contains(event.relatedTarget)) {
            return;
        }
        event.preventDefault();
        event.stopPropagation();
        this.isDragOver.set(false);
        this.dragOver.emit(false);
    }
    onDrop(event) {
        if (this.state.disabled()) {
            return;
        }
        event.preventDefault();
        this.isDragOver.set(false);
        this.dragOver.emit(false);
        const fileList = event.dataTransfer?.files;
        if (fileList) {
            const filteredFiles = fileDropFilter(fileList, this.state.fileTypes(), this.state.multiple());
            if (filteredFiles) {
                this.selected.emit(filteredFiles);
            }
            else {
                this.rejected.emit();
            }
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpFileDropzone, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "19.2.11", type: NgpFileDropzone, isStandalone: true, selector: "[ngpFileDropzone]", inputs: { fileTypes: { classPropertyName: "fileTypes", publicName: "ngpFileDropzoneFileTypes", isSignal: true, isRequired: false, transformFunction: null }, multiple: { classPropertyName: "multiple", publicName: "ngpFileDropzoneMultiple", isSignal: true, isRequired: false, transformFunction: null }, directory: { classPropertyName: "directory", publicName: "ngpFileDropzoneDirectory", isSignal: true, isRequired: false, transformFunction: null }, disabled: { classPropertyName: "disabled", publicName: "ngpFileDropzoneDisabled", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { selected: "ngpFileDropzoneSelected", rejected: "ngpFileDropzoneRejected", dragOver: "ngpFileDropzoneDragOver" }, host: { listeners: { "dragenter": "onDragEnter($event)", "dragover": "onDragOver($event)", "dragleave": "onDragLeave($event)", "drop": "onDrop($event)" }, properties: { "attr.data-dragover": "isDragOver() ? \"\" : null", "attr.data-disabled": "disabled() ? \"\" : null" } }, providers: [provideFileDropzoneState()], exportAs: ["ngpFileDropzone"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpFileDropzone, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpFileDropzone]',
                    exportAs: 'ngpFileDropzone',
                    providers: [provideFileDropzoneState()],
                    host: {
                        '[attr.data-dragover]': 'isDragOver() ? "" : null',
                        '[attr.data-disabled]': 'disabled() ? "" : null',
                    },
                }]
        }], ctorParameters: () => [], propDecorators: { onDragEnter: [{
                type: HostListener,
                args: ['dragenter', ['$event']]
            }], onDragOver: [{
                type: HostListener,
                args: ['dragover', ['$event']]
            }], onDragLeave: [{
                type: HostListener,
                args: ['dragleave', ['$event']]
            }], onDrop: [{
                type: HostListener,
                args: ['drop', ['$event']]
            }] } });

/**
 * The state token  for the FileUpload primitive.
 */
const NgpFileUploadStateToken = createStateToken('FileUpload');
/**
 * Provides the FileUpload state.
 */
const provideFileUploadState = createStateProvider(NgpFileUploadStateToken);
/**
 * Injects the FileUpload state.
 */
const injectFileUploadState = createStateInjector(NgpFileUploadStateToken);
/**
 * The FileUpload state registration function.
 */
const fileUploadState = createState(NgpFileUploadStateToken);

/**
 * A directive that allows you to turn any element into a file upload trigger.
 */
class NgpFileUpload {
    constructor() {
        /**
         * Access the document
         */
        this.document = inject(DOCUMENT);
        /**
         * Access the host element.
         */
        this.elementRef = inject(ElementRef);
        /**
         * The accepted file types. This can be an array of strings or a comma-separated string.
         * Accepted types can either be file extensions (e.g. `.jpg`) or MIME types (e.g. `image/jpeg`).
         */
        this.fileTypes = input(undefined, {
            alias: 'ngpFileUploadFileTypes',
            transform: types => coerceStringArray(types, ','),
        });
        /**
         * Whether to allow multiple files to be selected.
         */
        this.multiple = input(false, {
            alias: 'ngpFileUploadMultiple',
            transform: booleanAttribute,
        });
        /**
         * Whether to allow the user to select directories.
         */
        this.directory = input(false, {
            alias: 'ngpFileUploadDirectory',
            transform: booleanAttribute,
        });
        /**
         * Whether drag-and-drop is enabled.
         */
        this.dragAndDrop = input(true, {
            alias: 'ngpFileUploadDragDrop',
            transform: booleanAttribute,
        });
        /**
         * Whether the file upload is disabled.
         */
        this.disabled = input(false, {
            alias: 'ngpFileUploadDisabled',
            transform: booleanAttribute,
        });
        /**
         * Emits when the user selects files.
         */
        this.selected = output({
            alias: 'ngpFileUploadSelected',
        });
        /**
         * Emits when the user cancel the file selection.
         */
        this.canceled = output({
            alias: 'ngpFileUploadCanceled',
        });
        /**
         * Emits when uploaded files are rejected because they do not match the allowed {@link fileTypes}.
         */
        this.rejected = output({
            alias: 'ngpFileUploadRejected',
        });
        /**
         * Emits when the user drags a file over the file upload.
         */
        this.dragOver = output({
            alias: 'ngpFileUploadDragOver',
        });
        /**
         * Whether the user is currently dragging a file over the file upload.
         */
        this.isDragOver = signal(false);
        /**
         * Store the file input element.
         */
        this.input = this.document.createElement('input');
        /**
         * The file upload state.
         */
        this.state = fileUploadState(this);
        ngpInteractions({
            hover: true,
            press: true,
            focusVisible: true,
            disabled: this.state.disabled,
        });
        this.input.type = 'file';
        this.input.style.display = 'none';
        this.input.addEventListener('change', () => {
            this.selected.emit(this.input.files);
            // clear the input value to allow re-uploading the same file
            this.input.value = '';
        });
        this.input.addEventListener('cancel', () => this.canceled.emit());
    }
    showFileDialog() {
        if (this.state.disabled()) {
            return;
        }
        const fileTypes = this.state.fileTypes()?.join(',');
        if (fileTypes) {
            this.input.accept = fileTypes;
        }
        this.input.multiple = this.state.multiple();
        this.input.webkitdirectory = this.state.directory();
        this.input.click();
    }
    onDragEnter(event) {
        if (this.state.disabled() || !this.state.dragAndDrop()) {
            return;
        }
        event.preventDefault();
        event.stopPropagation();
        this.isDragOver.set(true);
        this.dragOver.emit(true);
    }
    onDragOver(event) {
        if (this.state.disabled() || !this.state.dragAndDrop()) {
            return;
        }
        event.stopPropagation();
        event.preventDefault();
        this.isDragOver.set(true);
    }
    onDragLeave(event) {
        if (this.state.disabled() || !this.state.dragAndDrop() || !this.isDragOver()) {
            return;
        }
        // if the element we are dragging over is a child of the file upload, ignore the event
        if (this.elementRef.nativeElement.contains(event.relatedTarget)) {
            return;
        }
        event.preventDefault();
        event.stopPropagation();
        this.isDragOver.set(false);
        this.dragOver.emit(false);
    }
    onDrop(event) {
        if (this.state.disabled() || !this.state.dragAndDrop()) {
            return;
        }
        event.preventDefault();
        this.isDragOver.set(false);
        this.dragOver.emit(false);
        const fileList = event.dataTransfer?.files;
        if (fileList) {
            const filteredFiles = fileDropFilter(fileList, this.state.fileTypes(), this.state.multiple());
            if (filteredFiles) {
                this.selected.emit(filteredFiles);
            }
            else {
                this.rejected.emit();
            }
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpFileUpload, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "19.2.11", type: NgpFileUpload, isStandalone: true, selector: "[ngpFileUpload]", inputs: { fileTypes: { classPropertyName: "fileTypes", publicName: "ngpFileUploadFileTypes", isSignal: true, isRequired: false, transformFunction: null }, multiple: { classPropertyName: "multiple", publicName: "ngpFileUploadMultiple", isSignal: true, isRequired: false, transformFunction: null }, directory: { classPropertyName: "directory", publicName: "ngpFileUploadDirectory", isSignal: true, isRequired: false, transformFunction: null }, dragAndDrop: { classPropertyName: "dragAndDrop", publicName: "ngpFileUploadDragDrop", isSignal: true, isRequired: false, transformFunction: null }, disabled: { classPropertyName: "disabled", publicName: "ngpFileUploadDisabled", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { selected: "ngpFileUploadSelected", canceled: "ngpFileUploadCanceled", rejected: "ngpFileUploadRejected", dragOver: "ngpFileUploadDragOver" }, host: { listeners: { "click": "showFileDialog()", "dragenter": "onDragEnter($event)", "dragover": "onDragOver($event)", "dragleave": "onDragLeave($event)", "drop": "onDrop($event)" }, properties: { "attr.data-disabled": "state.disabled() ? \"\" : null", "attr.data-dragover": "isDragOver() ? \"\" : null" } }, providers: [provideFileUploadState()], exportAs: ["ngpFileUpload"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpFileUpload, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpFileUpload]',
                    exportAs: 'ngpFileUpload',
                    providers: [provideFileUploadState()],
                    host: {
                        '[attr.data-disabled]': 'state.disabled() ? "" : null',
                        '[attr.data-dragover]': 'isDragOver() ? "" : null',
                    },
                }]
        }], ctorParameters: () => [], propDecorators: { showFileDialog: [{
                type: HostListener,
                args: ['click']
            }], onDragEnter: [{
                type: HostListener,
                args: ['dragenter', ['$event']]
            }], onDragOver: [{
                type: HostListener,
                args: ['dragover', ['$event']]
            }], onDragLeave: [{
                type: HostListener,
                args: ['dragleave', ['$event']]
            }], onDrop: [{
                type: HostListener,
                args: ['drop', ['$event']]
            }] } });

/**
 * Generated bundle index. Do not edit.
 */

export { NgpFileDropzone, NgpFileUpload, injectFileDropzoneState, injectFileUploadState, provideFileDropzoneState, provideFileUploadState };
