// ng-primitives/fesm2022/ng-primitives-resize.mjs
import * as i0 from '@angular/core';
import { inject, ElementRef, NgZone, DestroyRef, output, Directive } from '@angular/core';
import { fromResizeEvent } from 'ng-primitives/internal';
import { safeTakeUntilDestroyed } from 'ng-primitives/utils';

/**
 * Apply the `ngpResize` directive to an element to listen for resize events.
 */
class NgpResize {
    constructor() {
        /**
         * Access the element.
         */
        this.element = inject(ElementRef);
        /**
         * Access zone.js
         */
        this.ngZone = inject(NgZone);
        /**
         * Access the destroy ref
         */
        this.destroyRef = inject(DestroyRef);
        /**
         * Emits when the element is resized.
         */
        this.didResize = output({
            alias: 'ngpResize',
        });
        // observe the element for resize events
        fromResizeEvent(this.element.nativeElement)
            .pipe(safeTakeUntilDestroyed(this.destroyRef))
            .subscribe(event => this.ngZone.run(() => this.didResize.emit(event)));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpResize, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "19.2.11", type: NgpResize, isStandalone: true, selector: "[ngpResize]", outputs: { didResize: "ngpResize" }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpResize, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpResize]',
                }]
        }], ctorParameters: () => [] });

/**
 * Generated bundle index. Do not edit.
 */

export { NgpResize };
