// ng-primitives/fesm2022/ng-primitives-search.mjs
import * as i0 from '@angular/core';
import { HostListener, Directive, inject, DestroyRef, signal, computed } from '@angular/core';
import { createStateToken, createStateProvider, createStateInjector, createState } from 'ng-primitives/state';
import { safeTakeUntilDestroyed } from 'ng-primitives/utils';
import { fromEvent } from 'rxjs';

/**
 * The state token  for the Search primitive.
 */
const NgpSearchStateToken = createStateToken('Search');
/**
 * Provides the Search state.
 */
const provideSearchState = createStateProvider(NgpSearchStateToken);
/**
 * Injects the Search state.
 */
const injectSearchState = createStateInjector(NgpSearchStateToken);
/**
 * The Search state registration function.
 */
const searchState = createState(NgpSearchStateToken);

/**
 * The `NgpSearchClear` directive is can be added to a button to clear the search field on click.
 */
class NgpSearchClear {
    constructor() {
        /**
         * Access the Search instance.
         */
        this.search = injectSearchState();
    }
    /**
     * Clear the input field.
     */
    clear() {
        this.search().clear();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpSearchClear, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "19.2.11", type: NgpSearchClear, isStandalone: true, selector: "[ngpSearchClear]", host: { listeners: { "click": "clear()" }, properties: { "tabindex": "-1", "attr.data-empty": "search().empty() ? \"\" : null" } }, exportAs: ["ngpSearchClear"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpSearchClear, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpSearchClear]',
                    exportAs: 'ngpSearchClear',
                    host: {
                        '[tabindex]': '-1',
                        '[attr.data-empty]': 'search().empty() ? "" : null',
                    },
                }]
        }], propDecorators: { clear: [{
                type: HostListener,
                args: ['click']
            }] } });

/**
 * The `NgpSearch` directive is a container for the search field components.
 */
class NgpSearch {
    constructor() {
        /**
         * The destroy reference.
         */
        this.destroyRef = inject(DestroyRef);
        /**
         * The input field.
         */
        this.input = signal(null);
        /**
         * The value of the input.
         */
        this.value = signal('');
        /**
         * Whether the input field is empty.
         * @internal
         */
        this.empty = computed(() => this.value() === '');
        /**
         * The search field state.
         */
        this.state = searchState(this);
    }
    clear() {
        const input = this.input();
        if (!input) {
            return;
        }
        input.value = '';
        input.dispatchEvent(new Event('input', { bubbles: true }));
    }
    /**
     * Register the input field.
     * @param input The input field.
     * @internal
     */
    registerInput(input) {
        this.input.set(input);
        this.value.set(input.value);
        fromEvent(input, 'input')
            .pipe(safeTakeUntilDestroyed(this.destroyRef))
            .subscribe(() => this.value.set(input.value));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpSearch, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "19.2.11", type: NgpSearch, isStandalone: true, selector: "[ngpSearch]", host: { listeners: { "keydown.escape": "clear()" }, properties: { "attr.data-empty": "empty() ? \"\" : null" } }, providers: [provideSearchState()], exportAs: ["ngpSearch"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpSearch, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpSearch]',
                    exportAs: 'ngpSearch',
                    providers: [provideSearchState()],
                    host: {
                        '[attr.data-empty]': 'empty() ? "" : null',
                    },
                }]
        }], propDecorators: { clear: [{
                type: HostListener,
                args: ['keydown.escape']
            }] } });

/**
 * Generated bundle index. Do not edit.
 */

export { NgpSearch, NgpSearchClear, injectSearchState, provideSearchState };
