// ng-primitives/fesm2022/ng-primitives-toggle-group.mjs
import * as i0 from '@angular/core';
import { InjectionToken, inject, input, booleanAttribute, computed, Directive, output } from '@angular/core';
import * as i1 from 'ng-primitives/roving-focus';
import { NgpRovingFocusItem, injectRovingFocusGroupState, NgpRovingFocusGroup } from 'ng-primitives/roving-focus';
import { createStateToken, createStateProvider, createStateInjector, createState } from 'ng-primitives/state';
import { explicitEffect } from 'ng-primitives/internal';

const defaultToggleGroupConfig = {
    orientation: 'horizontal',
    type: 'single',
    allowDeselection: true,
};
const NgpToggleGroupConfigToken = new InjectionToken('NgpToggleGroupConfigToken');
/**
 * Provide the default ToggleGroup configuration
 * @param config The ToggleGroup configuration
 * @returns The provider
 */
function provideToggleGroupConfig(config) {
    return [
        {
            provide: NgpToggleGroupConfigToken,
            useValue: { ...defaultToggleGroupConfig, ...config },
        },
    ];
}
/**
 * Inject the ToggleGroup configuration
 * @returns The global ToggleGroup configuration
 */
function injectToggleGroupConfig() {
    return inject(NgpToggleGroupConfigToken, { optional: true }) ?? defaultToggleGroupConfig;
}

/**
 * The state token  for the ToggleGroup primitive.
 */
const NgpToggleGroupStateToken = createStateToken('ToggleGroup');
/**
 * Provides the ToggleGroup state.
 */
const provideToggleGroupState = createStateProvider(NgpToggleGroupStateToken);
/**
 * Injects the ToggleGroup state.
 */
const injectToggleGroupState = createStateInjector(NgpToggleGroupStateToken);
/**
 * The ToggleGroup state registration function.
 */
const toggleGroupState = createState(NgpToggleGroupStateToken);

/**
 * The state token  for the ToggleGroupItem primitive.
 */
const NgpToggleGroupItemStateToken = createStateToken('ToggleGroupItem');
/**
 * Provides the ToggleGroupItem state.
 */
const provideToggleGroupItemState = createStateProvider(NgpToggleGroupItemStateToken);
/**
 * Injects the ToggleGroupItem state.
 */
const injectToggleGroupItemState = createStateInjector(NgpToggleGroupItemStateToken);
/**
 * The ToggleGroupItem state registration function.
 */
const toggleGroupItemState = createState(NgpToggleGroupItemStateToken);

class NgpToggleGroupItem {
    constructor() {
        /**
         * Access the group that the item belongs to.
         */
        this.toggleGroup = injectToggleGroupState();
        /**
         * The value of the item.
         * @required
         */
        this.value = input(undefined, {
            alias: 'ngpToggleGroupItemValue',
        });
        /**
         * Whether the item is disabled.
         */
        this.disabled = input(false, {
            alias: 'ngpToggleGroupItemDisabled',
            transform: booleanAttribute,
        });
        /**
         * Whether the item is selected.
         */
        this.selected = computed(() => this.toggleGroup().isSelected(this.state.value()));
        /**
         * The state of the item.
         */
        this.state = toggleGroupItemState(this);
    }
    ngOnInit() {
        // we can't use a required input for value as it is used in a computed property before the input is set
        if (this.state.value() === undefined) {
            throw new Error('The value input is required for the toggle group item.');
        }
    }
    /**
     * Toggle the item.
     */
    toggle() {
        if (this.state.disabled()) {
            return;
        }
        this.toggleGroup().toggle(this.state.value());
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpToggleGroupItem, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "19.2.11", type: NgpToggleGroupItem, isStandalone: true, selector: "[ngpToggleGroupItem]", inputs: { value: { classPropertyName: "value", publicName: "ngpToggleGroupItemValue", isSignal: true, isRequired: false, transformFunction: null }, disabled: { classPropertyName: "disabled", publicName: "ngpToggleGroupItemDisabled", isSignal: true, isRequired: false, transformFunction: null } }, host: { attributes: { "role": "radio" }, listeners: { "click": "toggle()" }, properties: { "attr.aria-checked": "selected()", "attr.data-selected": "selected() ? \"\" : null", "attr.aria-disabled": "state.disabled()", "attr.data-disabled": "state.disabled() ? \"\" : null" } }, providers: [provideToggleGroupItemState()], exportAs: ["ngpToggleGroupItem"], hostDirectives: [{ directive: i1.NgpRovingFocusItem, inputs: ["ngpRovingFocusItemDisabled", "ngpToggleGroupItemDisabled"] }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpToggleGroupItem, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpToggleGroupItem]',
                    exportAs: 'ngpToggleGroupItem',
                    providers: [provideToggleGroupItemState()],
                    hostDirectives: [
                        {
                            directive: NgpRovingFocusItem,
                            inputs: ['ngpRovingFocusItemDisabled: ngpToggleGroupItemDisabled'],
                        },
                    ],
                    host: {
                        role: 'radio',
                        '[attr.aria-checked]': 'selected()',
                        '[attr.data-selected]': 'selected() ? "" : null',
                        '[attr.aria-disabled]': 'state.disabled()',
                        '[attr.data-disabled]': 'state.disabled() ? "" : null',
                        '(click)': 'toggle()',
                    },
                }]
        }] });

class NgpToggleGroup {
    constructor() {
        /**
         * Access the roving focus group state.
         */
        this.rovingFocusGroupState = injectRovingFocusGroupState();
        /**
         * Access the global toggle group configuration.
         */
        this.config = injectToggleGroupConfig();
        /**
         * The orientation of the toggle group.
         */
        this.orientation = input(this.config.orientation, {
            alias: 'ngpToggleGroupOrientation',
        });
        /**
         * Whether toggle buttons can be deselected. If set to `false`, clicking a selected toggle button will not deselect it.
         * @default true
         */
        this.allowDeselection = input(this.config.allowDeselection, {
            alias: 'ngpToggleGroupAllowDeselection',
            transform: booleanAttribute,
        });
        /**
         * The type of the toggle group, whether only one item can be selected or multiple.
         */
        this.type = input(this.config.type, { alias: 'ngpToggleGroupType' });
        /**
         * The selected value(s) of the toggle group.
         */
        this.value = input([], { alias: 'ngpToggleGroupValue' });
        /**
         * Emits when the value of the toggle group changes.
         */
        this.valueChange = output({ alias: 'ngpToggleGroupValueChange' });
        /**
         * Whether the toggle group is disabled.
         */
        this.disabled = input(false, {
            alias: 'ngpToggleGroupDisabled',
            transform: booleanAttribute,
        });
        /**
         * The state of the toggle group.
         */
        this.state = toggleGroupState(this);
        // the roving focus group defaults to vertical orientation whereas
        // the default for the toggle group may be different if provided via global config
        explicitEffect([this.state.orientation], ([orientation]) => this.rovingFocusGroupState().orientation.set(orientation));
    }
    /**
     * Select a value in the toggle group.
     */
    select(value) {
        if (this.state.disabled()) {
            return;
        }
        let newValue = [];
        if (this.state.type() === 'single') {
            newValue = [value];
        }
        else {
            newValue = [...this.state.value(), value];
        }
        this.state.value.set(newValue);
        this.valueChange.emit(newValue);
    }
    /**
     * De-select a value in the toggle group.
     */
    deselect(value) {
        if (this.state.disabled() || !this.state.allowDeselection()) {
            return;
        }
        const newValue = this.state.value().filter(v => v !== value);
        this.state.value.set(newValue);
        this.valueChange.emit(newValue);
    }
    /**
     * Check if a value is selected in the toggle group.
     * @internal
     */
    isSelected(value) {
        return this.state.value()?.includes(value) ?? false;
    }
    /**
     * Toggle a value in the toggle group.
     * @internal
     */
    toggle(value) {
        if (this.isSelected(value)) {
            this.deselect(value);
        }
        else {
            this.select(value);
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpToggleGroup, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "19.2.11", type: NgpToggleGroup, isStandalone: true, selector: "[ngpToggleGroup]", inputs: { orientation: { classPropertyName: "orientation", publicName: "ngpToggleGroupOrientation", isSignal: true, isRequired: false, transformFunction: null }, allowDeselection: { classPropertyName: "allowDeselection", publicName: "ngpToggleGroupAllowDeselection", isSignal: true, isRequired: false, transformFunction: null }, type: { classPropertyName: "type", publicName: "ngpToggleGroupType", isSignal: true, isRequired: false, transformFunction: null }, value: { classPropertyName: "value", publicName: "ngpToggleGroupValue", isSignal: true, isRequired: false, transformFunction: null }, disabled: { classPropertyName: "disabled", publicName: "ngpToggleGroupDisabled", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { valueChange: "ngpToggleGroupValueChange" }, host: { attributes: { "role": "group" }, properties: { "attr.data-orientation": "state.orientation()", "attr.data-type": "state.type()", "attr.data-disabled": "state.disabled() ? \"\" : null" } }, providers: [provideToggleGroupState()], exportAs: ["ngpToggleGroup"], hostDirectives: [{ directive: i1.NgpRovingFocusGroup }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.11", ngImport: i0, type: NgpToggleGroup, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpToggleGroup]',
                    exportAs: 'ngpToggleGroup',
                    providers: [provideToggleGroupState()],
                    hostDirectives: [NgpRovingFocusGroup],
                    host: {
                        role: 'group',
                        '[attr.data-orientation]': 'state.orientation()',
                        '[attr.data-type]': 'state.type()',
                        '[attr.data-disabled]': 'state.disabled() ? "" : null',
                    },
                }]
        }], ctorParameters: () => [] });

/**
 * Generated bundle index. Do not edit.
 */

export { NgpToggleGroup, NgpToggleGroupItem, injectToggleGroupItemState, injectToggleGroupState, provideToggleGroupConfig, provideToggleGroupItemState, provideToggleGroupState };
